// @aws-sdk/client-cost-explorer/dist-cjs/index.js
'use strict';

var middlewareHostHeader = require('@aws-sdk/middleware-host-header');
var middlewareLogger = require('@aws-sdk/middleware-logger');
var middlewareRecursionDetection = require('@aws-sdk/middleware-recursion-detection');
var middlewareUserAgent = require('@aws-sdk/middleware-user-agent');
var configResolver = require('@smithy/config-resolver');
var core = require('@smithy/core');
var schema = require('@smithy/core/schema');
var middlewareContentLength = require('@smithy/middleware-content-length');
var middlewareEndpoint = require('@smithy/middleware-endpoint');
var middlewareRetry = require('@smithy/middleware-retry');
var smithyClient = require('@smithy/smithy-client');
var httpAuthSchemeProvider = require('./auth/httpAuthSchemeProvider');
var runtimeConfig = require('./runtimeConfig');
var regionConfigResolver = require('@aws-sdk/region-config-resolver');
var protocolHttp = require('@smithy/protocol-http');

const resolveClientEndpointParameters = (options) => {
    return Object.assign(options, {
        useDualstackEndpoint: options.useDualstackEndpoint ?? false,
        useFipsEndpoint: options.useFipsEndpoint ?? false,
        defaultSigningName: "ce",
    });
};
const commonParams = {
    UseFIPS: { type: "builtInParams", name: "useFipsEndpoint" },
    Endpoint: { type: "builtInParams", name: "endpoint" },
    Region: { type: "builtInParams", name: "region" },
    UseDualStack: { type: "builtInParams", name: "useDualstackEndpoint" },
};

const getHttpAuthExtensionConfiguration = (runtimeConfig) => {
    const _httpAuthSchemes = runtimeConfig.httpAuthSchemes;
    let _httpAuthSchemeProvider = runtimeConfig.httpAuthSchemeProvider;
    let _credentials = runtimeConfig.credentials;
    return {
        setHttpAuthScheme(httpAuthScheme) {
            const index = _httpAuthSchemes.findIndex((scheme) => scheme.schemeId === httpAuthScheme.schemeId);
            if (index === -1) {
                _httpAuthSchemes.push(httpAuthScheme);
            }
            else {
                _httpAuthSchemes.splice(index, 1, httpAuthScheme);
            }
        },
        httpAuthSchemes() {
            return _httpAuthSchemes;
        },
        setHttpAuthSchemeProvider(httpAuthSchemeProvider) {
            _httpAuthSchemeProvider = httpAuthSchemeProvider;
        },
        httpAuthSchemeProvider() {
            return _httpAuthSchemeProvider;
        },
        setCredentials(credentials) {
            _credentials = credentials;
        },
        credentials() {
            return _credentials;
        },
    };
};
const resolveHttpAuthRuntimeConfig = (config) => {
    return {
        httpAuthSchemes: config.httpAuthSchemes(),
        httpAuthSchemeProvider: config.httpAuthSchemeProvider(),
        credentials: config.credentials(),
    };
};

const resolveRuntimeExtensions = (runtimeConfig, extensions) => {
    const extensionConfiguration = Object.assign(regionConfigResolver.getAwsRegionExtensionConfiguration(runtimeConfig), smithyClient.getDefaultExtensionConfiguration(runtimeConfig), protocolHttp.getHttpHandlerExtensionConfiguration(runtimeConfig), getHttpAuthExtensionConfiguration(runtimeConfig));
    extensions.forEach((extension) => extension.configure(extensionConfiguration));
    return Object.assign(runtimeConfig, regionConfigResolver.resolveAwsRegionExtensionConfiguration(extensionConfiguration), smithyClient.resolveDefaultRuntimeConfig(extensionConfiguration), protocolHttp.resolveHttpHandlerRuntimeConfig(extensionConfiguration), resolveHttpAuthRuntimeConfig(extensionConfiguration));
};

class CostExplorerClient extends smithyClient.Client {
    config;
    constructor(...[configuration]) {
        const _config_0 = runtimeConfig.getRuntimeConfig(configuration || {});
        super(_config_0);
        this.initConfig = _config_0;
        const _config_1 = resolveClientEndpointParameters(_config_0);
        const _config_2 = middlewareUserAgent.resolveUserAgentConfig(_config_1);
        const _config_3 = middlewareRetry.resolveRetryConfig(_config_2);
        const _config_4 = configResolver.resolveRegionConfig(_config_3);
        const _config_5 = middlewareHostHeader.resolveHostHeaderConfig(_config_4);
        const _config_6 = middlewareEndpoint.resolveEndpointConfig(_config_5);
        const _config_7 = httpAuthSchemeProvider.resolveHttpAuthSchemeConfig(_config_6);
        const _config_8 = resolveRuntimeExtensions(_config_7, configuration?.extensions || []);
        this.config = _config_8;
        this.middlewareStack.use(schema.getSchemaSerdePlugin(this.config));
        this.middlewareStack.use(middlewareUserAgent.getUserAgentPlugin(this.config));
        this.middlewareStack.use(middlewareRetry.getRetryPlugin(this.config));
        this.middlewareStack.use(middlewareContentLength.getContentLengthPlugin(this.config));
        this.middlewareStack.use(middlewareHostHeader.getHostHeaderPlugin(this.config));
        this.middlewareStack.use(middlewareLogger.getLoggerPlugin(this.config));
        this.middlewareStack.use(middlewareRecursionDetection.getRecursionDetectionPlugin(this.config));
        this.middlewareStack.use(core.getHttpAuthSchemeEndpointRuleSetPlugin(this.config, {
            httpAuthSchemeParametersProvider: httpAuthSchemeProvider.defaultCostExplorerHttpAuthSchemeParametersProvider,
            identityProviderConfigProvider: async (config) => new core.DefaultIdentityProviderConfig({
                "aws.auth#sigv4": config.credentials,
            }),
        }));
        this.middlewareStack.use(core.getHttpSigningPlugin(this.config));
    }
    destroy() {
        super.destroy();
    }
}

let CostExplorerServiceException$1 = class CostExplorerServiceException extends smithyClient.ServiceException {
    constructor(options) {
        super(options);
        Object.setPrototypeOf(this, CostExplorerServiceException.prototype);
    }
};

let AnalysisNotFoundException$1 = class AnalysisNotFoundException extends CostExplorerServiceException$1 {
    name = "AnalysisNotFoundException";
    $fault = "client";
    Message;
    constructor(opts) {
        super({
            name: "AnalysisNotFoundException",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, AnalysisNotFoundException.prototype);
        this.Message = opts.Message;
    }
};
let LimitExceededException$1 = class LimitExceededException extends CostExplorerServiceException$1 {
    name = "LimitExceededException";
    $fault = "client";
    Message;
    constructor(opts) {
        super({
            name: "LimitExceededException",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, LimitExceededException.prototype);
        this.Message = opts.Message;
    }
};
let UnknownMonitorException$1 = class UnknownMonitorException extends CostExplorerServiceException$1 {
    name = "UnknownMonitorException";
    $fault = "client";
    Message;
    constructor(opts) {
        super({
            name: "UnknownMonitorException",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, UnknownMonitorException.prototype);
        this.Message = opts.Message;
    }
};
let ServiceQuotaExceededException$1 = class ServiceQuotaExceededException extends CostExplorerServiceException$1 {
    name = "ServiceQuotaExceededException";
    $fault = "client";
    Message;
    constructor(opts) {
        super({
            name: "ServiceQuotaExceededException",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, ServiceQuotaExceededException.prototype);
        this.Message = opts.Message;
    }
};
let UnknownSubscriptionException$1 = class UnknownSubscriptionException extends CostExplorerServiceException$1 {
    name = "UnknownSubscriptionException";
    $fault = "client";
    Message;
    constructor(opts) {
        super({
            name: "UnknownSubscriptionException",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, UnknownSubscriptionException.prototype);
        this.Message = opts.Message;
    }
};
let ResourceNotFoundException$1 = class ResourceNotFoundException extends CostExplorerServiceException$1 {
    name = "ResourceNotFoundException";
    $fault = "client";
    Message;
    ResourceName;
    constructor(opts) {
        super({
            name: "ResourceNotFoundException",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, ResourceNotFoundException.prototype);
        this.Message = opts.Message;
        this.ResourceName = opts.ResourceName;
    }
};
let InvalidNextTokenException$1 = class InvalidNextTokenException extends CostExplorerServiceException$1 {
    name = "InvalidNextTokenException";
    $fault = "client";
    Message;
    constructor(opts) {
        super({
            name: "InvalidNextTokenException",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, InvalidNextTokenException.prototype);
        this.Message = opts.Message;
    }
};
let DataUnavailableException$1 = class DataUnavailableException extends CostExplorerServiceException$1 {
    name = "DataUnavailableException";
    $fault = "client";
    Message;
    constructor(opts) {
        super({
            name: "DataUnavailableException",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, DataUnavailableException.prototype);
        this.Message = opts.Message;
    }
};
let BillExpirationException$1 = class BillExpirationException extends CostExplorerServiceException$1 {
    name = "BillExpirationException";
    $fault = "client";
    Message;
    constructor(opts) {
        super({
            name: "BillExpirationException",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, BillExpirationException.prototype);
        this.Message = opts.Message;
    }
};
let BillingViewHealthStatusException$1 = class BillingViewHealthStatusException extends CostExplorerServiceException$1 {
    name = "BillingViewHealthStatusException";
    $fault = "client";
    Message;
    constructor(opts) {
        super({
            name: "BillingViewHealthStatusException",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, BillingViewHealthStatusException.prototype);
        this.Message = opts.Message;
    }
};
let RequestChangedException$1 = class RequestChangedException extends CostExplorerServiceException$1 {
    name = "RequestChangedException";
    $fault = "client";
    Message;
    constructor(opts) {
        super({
            name: "RequestChangedException",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, RequestChangedException.prototype);
        this.Message = opts.Message;
    }
};
let UnresolvableUsageUnitException$1 = class UnresolvableUsageUnitException extends CostExplorerServiceException$1 {
    name = "UnresolvableUsageUnitException";
    $fault = "client";
    Message;
    constructor(opts) {
        super({
            name: "UnresolvableUsageUnitException",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, UnresolvableUsageUnitException.prototype);
        this.Message = opts.Message;
    }
};
let GenerationExistsException$1 = class GenerationExistsException extends CostExplorerServiceException$1 {
    name = "GenerationExistsException";
    $fault = "client";
    Message;
    constructor(opts) {
        super({
            name: "GenerationExistsException",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, GenerationExistsException.prototype);
        this.Message = opts.Message;
    }
};
let BackfillLimitExceededException$1 = class BackfillLimitExceededException extends CostExplorerServiceException$1 {
    name = "BackfillLimitExceededException";
    $fault = "client";
    Message;
    constructor(opts) {
        super({
            name: "BackfillLimitExceededException",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, BackfillLimitExceededException.prototype);
        this.Message = opts.Message;
    }
};
let TooManyTagsException$1 = class TooManyTagsException extends CostExplorerServiceException$1 {
    name = "TooManyTagsException";
    $fault = "client";
    Message;
    ResourceName;
    constructor(opts) {
        super({
            name: "TooManyTagsException",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, TooManyTagsException.prototype);
        this.Message = opts.Message;
        this.ResourceName = opts.ResourceName;
    }
};

const _A = "Anomaly";
const _AC = "AmortizedCommitment";
const _ACT = "AnalysisCompletionTime";
const _AD = "AnalysisDetails";
const _ADI = "AnomalyDateInterval";
const _ADp = "ApproximationDimension";
const _AED = "AnomalyEndDate";
const _AI = "AnalysisId";
const _AIc = "AccountId";
const _AIn = "AnomalyId";
const _AIna = "AnalysisIds";
const _AM = "AnomalyMonitor";
const _AMd = "AdditionalMetadata";
const _AMn = "AnomalyMonitors";
const _ANFE = "AnalysisNotFoundException";
const _ANOCUUPH = "AverageNumberOfCapacityUnitsUsedPerHour";
const _ANOIUPH = "AverageNumberOfInstancesUsedPerHour";
const _ANUUPH = "AverageNormalizedUnitsUsedPerHour";
const _ARC = "AmortizedRecurringCommitment";
const _ARF = "AmortizedRecurringFee";
const _AS = "AnalysisSummary";
const _ASD = "AnomalyStartDate";
const _ASL = "AnalysisSummaryList";
const _AST = "AnalysisStartedTime";
const _ASc = "AccountScope";
const _ASn = "AnalysisStatus";
const _ASno = "AnomalyScore";
const _ASnom = "AnomalySubscription";
const _ASnoma = "AnomalySubscriptions";
const _AT = "AnalysisType";
const _AU = "AverageUtilization";
const _AUC = "AmortizedUpfrontCommitment";
const _AUF = "AmortizedUpfrontFee";
const _AZ = "AvailabilityZone";
const _Ad = "Address";
const _Am = "Amount";
const _An = "And";
const _Ano = "Anomalies";
const _At = "Attributes";
const _BC = "BenefitsConsidered";
const _BEE = "BillExpirationException";
const _BF = "BackfillFrom";
const _BLEE = "BackfillLimitExceededException";
const _BR = "BackfillRequests";
const _BRa = "BackfillRequest";
const _BS = "BackfillStatus";
const _BTP = "BaselineTimePeriod";
const _BTPA = "BaselineTimePeriodAmount";
const _BVA = "BillingViewArn";
const _BVHSE = "BillingViewHealthStatusException";
const _C = "Component";
const _CA = "CompletedAt";
const _CAC = "CurrentAverageCoverage";
const _CAHODS = "CurrentAverageHourlyOnDemandSpend";
const _CAM = "CreateAnomalyMonitor";
const _CAMR = "CreateAnomalyMonitorRequest";
const _CAMRr = "CreateAnomalyMonitorResponse";
const _CAS = "CreateAnomalySubscription";
const _CASR = "CreateAnomalySubscriptionRequest";
const _CASRr = "CreateAnomalySubscriptionResponse";
const _CAT = "CostAllocationTag";
const _CATBR = "CostAllocationTagBackfillRequest";
const _CATBRL = "CostAllocationTagBackfillRequestList";
const _CATL = "CostAllocationTagList";
const _CATS = "CostAllocationTagsStatus";
const _CATSE = "CostAllocationTagStatusEntry";
const _CATSL = "CostAllocationTagStatusList";
const _CATo = "CostAllocationTags";
const _CAUC = "CostAndUsageComparison";
const _CAUCo = "CostAndUsageComparisons";
const _CAUS = "CostAndUsageSelector";
const _CBT = "CoverageByTime";
const _CBTo = "CoveragesByTime";
const _CC = "CostCategory";
const _CCA = "CostCategoryArn";
const _CCCD = "CreateCostCategoryDefinition";
const _CCCDR = "CreateCostCategoryDefinitionRequest";
const _CCCDRr = "CreateCostCategoryDefinitionResponse";
const _CCD = "CostComparisonDriver";
const _CCDo = "CostComparisonDrivers";
const _CCIVD = "CostCategoryInheritedValueDimension";
const _CCN = "CostCategoryName";
const _CCNo = "CostCategoryNames";
const _CCPS = "CostCategoryProcessingStatus";
const _CCPSL = "CostCategoryProcessingStatusList";
const _CCR = "CostCategoryReference";
const _CCRL = "CostCategoryReferencesList";
const _CCRLo = "CostCategoryRulesList";
const _CCRo = "CostCategoryRule";
const _CCRos = "CostCategoryReferences";
const _CCSCR = "CostCategorySplitChargeRule";
const _CCSCRL = "CostCategorySplitChargeRulesList";
const _CCSCRP = "CostCategorySplitChargeRuleParameter";
const _CCSCRPL = "CostCategorySplitChargeRuleParametersList";
const _CCV = "CostCategoryValues";
const _CCo = "CoverageCost";
const _CCos = "CostCategories";
const _CCu = "CurrencyCode";
const _CCur = "CurrentCoverage";
const _CD = "CreationDate";
const _CDo = "CostDrivers";
const _CDos = "CostDriver";
const _CG = "CurrentGeneration";
const _CH = "CoverageHours";
const _CHP = "CoverageHoursPercentage";
const _CI = "CurrentInstance";
const _CM = "ComparisonMetrics";
const _CMHODS = "CurrentMaximumHourlyOnDemandSpend";
const _CMHODSu = "CurrentMinimumHourlyOnDemandSpend";
const _CMV = "ComparisonMetricValue";
const _CNU = "CoverageNormalizedUnits";
const _CNUP = "CoverageNormalizedUnitsPercentage";
const _CODS = "CurrentOnDemandSpend";
const _CP = "CoveragePercentage";
const _CPAC = "CommitmentPurchaseAnalysisConfiguration";
const _CS = "CurrentScore";
const _CSo = "CostSelector";
const _CTP = "ComparisonTimePeriod";
const _CTPA = "ComparisonTimePeriodAmount";
const _CU = "CapacityUnits";
const _Co = "Coverage";
const _Cod = "Code";
const _Con = "Context";
const _Conf = "Configuration";
const _Cont = "Contribution";
const _D = "Difference";
const _DAM = "DeleteAnomalyMonitor";
const _DAMR = "DeleteAnomalyMonitorRequest";
const _DAMRe = "DeleteAnomalyMonitorResponse";
const _DAS = "DeleteAnomalySubscription";
const _DASR = "DeleteAnomalySubscriptionRequest";
const _DASRe = "DeleteAnomalySubscriptionResponse";
const _DCCD = "DeleteCostCategoryDefinition";
const _DCCDR = "DeleteCostCategoryDefinitionRequest";
const _DCCDRe = "DeleteCostCategoryDefinitionResponse";
const _DCCDRes = "DescribeCostCategoryDefinitionRequest";
const _DCCDResc = "DescribeCostCategoryDefinitionResponse";
const _DCCDe = "DescribeCostCategoryDefinition";
const _DCTP = "DailyCommitmentToPurchase";
const _DDBCD = "DynamoDBCapacityDetails";
const _DE = "DatabaseEngine";
const _DEa = "DatabaseEdition";
const _DI = "DateInterval";
const _DK = "DimensionKey";
const _DN = "DimensionName";
const _DO = "DeploymentOption";
const _DRBPS = "DiskReadBytesPerSecond";
const _DROPS = "DiskReadOpsPerSecond";
const _DRU = "DiskResourceUtilization";
const _DT = "DataType";
const _DTI = "DefaultTargetInstance";
const _DUE = "DataUnavailableException";
const _DV = "DimensionValue";
const _DVA = "DimensionValueAttributes";
const _DVC = "DimensionalValueCount";
const _DVWA = "DimensionValuesWithAttributes";
const _DVWAL = "DimensionValuesWithAttributesList";
const _DVe = "DefaultValue";
const _DVi = "DimensionValues";
const _DWBPS = "DiskWriteBytesPerSecond";
const _DWOPS = "DiskWriteOpsPerSecond";
const _Di = "Dimensions";
const _Dim = "Dimension";
const _E = "End";
const _EAC = "EstimatedAverageCoverage";
const _EAU = "EstimatedAverageUtilization";
const _EBEIM = "EstimatedBreakEvenInMonths";
const _EBSRU = "EBSResourceUtilization";
const _EC = "ErrorCode";
const _ECC = "EstimatedCommitmentCost";
const _ECID = "EC2InstanceDetails";
const _ECIDl = "ElastiCacheInstanceDetails";
const _ECRD = "EC2ResourceDetails";
const _ECRU = "EC2ResourceUtilization";
const _ECS = "EC2Specification";
const _ECT = "EstimatedCompletionTime";
const _ECs = "EstimatedCoverage";
const _ED = "EndDate";
const _EE = "EffectiveEnd";
const _EHC = "ExistingHourlyCommitment";
const _EMC = "EstimatedMonthlyCost";
const _EMODC = "EstimatedMonthlyOnDemandCost";
const _EMS = "EstimatedMonthlySavings";
const _EMSA = "EstimatedMonthlySavingsAmount";
const _EMSP = "EstimatedMonthlySavingsPercentage";
const _ENCU = "EstimatedNewCommitmentUtilization";
const _EO = "EffectiveOn";
const _EODC = "EstimatedOnDemandCost";
const _EODCWCC = "EstimatedOnDemandCostWithCurrentCommitment";
const _ERBPS = "EbsReadBytesPerSecond";
const _ERCFLP = "EstimatedReservationCostForLookbackPeriod";
const _EROI = "EstimatedROI";
const _EROPS = "EbsReadOpsPerSecond";
const _ERU = "ExpectedResourceUtilization";
const _ES = "EffectiveStart";
const _ESA = "EstimatedSavingsAmount";
const _ESID = "ESInstanceDetails";
const _ESP = "EstimatedSavingsPercentage";
const _ESPC = "EstimatedSPCost";
const _ETC = "EstimatedTotalCost";
const _ETMSA = "EstimatedTotalMonthlySavingsAmount";
const _EV = "EndValue";
const _EWBPS = "EbsWriteBytesPerSecond";
const _EWOPS = "EbsWriteOpsPerSecond";
const _Er = "Errors";
const _Es = "Estimated";
const _Ex = "Expression";
const _Exp = "Expressions";
const _F = "Feedback";
const _FR = "ForecastResult";
const _FRBT = "ForecastResultsByTime";
const _FRC = "FindingReasonCodes";
const _Fa = "Family";
const _Fi = "Filter";
const _Fr = "Frequency";
const _G = "Groups";
const _GA = "GetAnomalies";
const _GAM = "GetAnomalyMonitors";
const _GAMR = "GetAnomalyMonitorsRequest";
const _GAMRe = "GetAnomalyMonitorsResponse";
const _GAR = "GetAnomaliesRequest";
const _GARe = "GetAnomaliesResponse";
const _GAS = "GetAnomalySubscriptions";
const _GASR = "GetAnomalySubscriptionsRequest";
const _GASRe = "GetAnomalySubscriptionsResponse";
const _GAUR = "GetApproximateUsageRecords";
const _GAURR = "GetApproximateUsageRecordsRequest";
const _GAURRe = "GetApproximateUsageRecordsResponse";
const _GB = "GroupBy";
const _GCAU = "GetCostAndUsage";
const _GCAUC = "GetCostAndUsageComparisons";
const _GCAUCR = "GetCostAndUsageComparisonsRequest";
const _GCAUCRe = "GetCostAndUsageComparisonsResponse";
const _GCAUR = "GetCostAndUsageRequest";
const _GCAURe = "GetCostAndUsageResponse";
const _GCAUWR = "GetCostAndUsageWithResources";
const _GCAUWRR = "GetCostAndUsageWithResourcesRequest";
const _GCAUWRRe = "GetCostAndUsageWithResourcesResponse";
const _GCC = "GetCostCategories";
const _GCCD = "GetCostComparisonDrivers";
const _GCCDR = "GetCostComparisonDriversRequest";
const _GCCDRe = "GetCostComparisonDriversResponse";
const _GCCR = "GetCostCategoriesRequest";
const _GCCRe = "GetCostCategoriesResponse";
const _GCF = "GetCostForecast";
const _GCFR = "GetCostForecastRequest";
const _GCFRe = "GetCostForecastResponse";
const _GCPA = "GetCommitmentPurchaseAnalysis";
const _GCPAR = "GetCommitmentPurchaseAnalysisRequest";
const _GCPARe = "GetCommitmentPurchaseAnalysisResponse";
const _GCT = "GenerationCompletionTime";
const _GD = "GroupDefinitions";
const _GDV = "GetDimensionValues";
const _GDVR = "GetDimensionValuesRequest";
const _GDVRe = "GetDimensionValuesResponse";
const _GDr = "GroupDefinition";
const _GEE = "GenerationExistsException";
const _GRC = "GetReservationCoverage";
const _GRCR = "GetReservationCoverageRequest";
const _GRCRe = "GetReservationCoverageResponse";
const _GRPR = "GetReservationPurchaseRecommendation";
const _GRPRR = "GetReservationPurchaseRecommendationRequest";
const _GRPRRe = "GetReservationPurchaseRecommendationResponse";
const _GRR = "GetRightsizingRecommendation";
const _GRRR = "GetRightsizingRecommendationRequest";
const _GRRRe = "GetRightsizingRecommendationResponse";
const _GRU = "GetReservationUtilization";
const _GRUR = "GetReservationUtilizationRequest";
const _GRURe = "GetReservationUtilizationResponse";
const _GS = "GenerationSummary";
const _GSL = "GenerationSummaryList";
const _GSPC = "GetSavingsPlansCoverage";
const _GSPCR = "GetSavingsPlansCoverageRequest";
const _GSPCRe = "GetSavingsPlansCoverageResponse";
const _GSPPR = "GetSavingsPlansPurchaseRecommendation";
const _GSPPRD = "GetSavingsPlanPurchaseRecommendationDetails";
const _GSPPRDR = "GetSavingsPlanPurchaseRecommendationDetailsRequest";
const _GSPPRDRe = "GetSavingsPlanPurchaseRecommendationDetailsResponse";
const _GSPPRR = "GetSavingsPlansPurchaseRecommendationRequest";
const _GSPPRRe = "GetSavingsPlansPurchaseRecommendationResponse";
const _GSPU = "GetSavingsPlansUtilization";
const _GSPUD = "GetSavingsPlansUtilizationDetails";
const _GSPUDR = "GetSavingsPlansUtilizationDetailsRequest";
const _GSPUDRe = "GetSavingsPlansUtilizationDetailsResponse";
const _GSPUR = "GetSavingsPlansUtilizationRequest";
const _GSPURe = "GetSavingsPlansUtilizationResponse";
const _GST = "GenerationStartedTime";
const _GSe = "GenerationStatus";
const _GT = "GenerationTimestamp";
const _GTR = "GetTagsRequest";
const _GTRe = "GetTagsResponse";
const _GTe = "GetTags";
const _GUF = "GetUsageForecast";
const _GUFR = "GetUsageForecastRequest";
const _GUFRe = "GetUsageForecastResponse";
const _Gr = "Granularity";
const _Gro = "Group";
const _HCTP = "HourlyCommitmentToPurchase";
const _HODR = "HourlyOnDemandRate";
const _I = "Impact";
const _IC = "InstanceClass";
const _ID = "InstanceDetails";
const _IF = "InstanceFamily";
const _IN = "InstanceName";
const _INTE = "InvalidNextTokenException";
const _IS = "InstanceSize";
const _IT = "InstanceType";
const _IV = "InheritedValue";
const _K = "Key";
const _Ke = "Keys";
const _LA = "LinkedAccount";
const _LAN = "LinkedAccountName";
const _LBTP = "LookBackTimePeriod";
const _LCAT = "ListCostAllocationTags";
const _LCATBH = "ListCostAllocationTagBackfillHistory";
const _LCATBHR = "ListCostAllocationTagBackfillHistoryRequest";
const _LCATBHRi = "ListCostAllocationTagBackfillHistoryResponse";
const _LCATR = "ListCostAllocationTagsRequest";
const _LCATRi = "ListCostAllocationTagsResponse";
const _LCCD = "ListCostCategoryDefinitions";
const _LCCDR = "ListCostCategoryDefinitionsRequest";
const _LCCDRi = "ListCostCategoryDefinitionsResponse";
const _LCPA = "ListCommitmentPurchaseAnalyses";
const _LCPAR = "ListCommitmentPurchaseAnalysesRequest";
const _LCPARi = "ListCommitmentPurchaseAnalysesResponse";
const _LED = "LastEvaluatedDate";
const _LEE = "LimitExceededException";
const _LM = "LicenseModel";
const _LP = "LookbackPeriod";
const _LPID = "LookbackPeriodInDays";
const _LPIH = "LookbackPeriodInHours";
const _LSPPRG = "ListSavingsPlansPurchaseRecommendationGeneration";
const _LSPPRGR = "ListSavingsPlansPurchaseRecommendationGenerationRequest";
const _LSPPRGRi = "ListSavingsPlansPurchaseRecommendationGenerationResponse";
const _LTFR = "ListTagsForResource";
const _LTFRR = "ListTagsForResourceRequest";
const _LTFRRi = "ListTagsForResourceResponse";
const _LUA = "LastUpdatedAt";
const _LUD = "LastUpdatedDate";
const _LUDa = "LastUsedDate";
const _LUT = "LatestUsageTimestamp";
const _M = "Message";
const _MA = "MonitorArn";
const _MAL = "MonitorArnList";
const _MC = "MonthlyCost";
const _MCUP = "MaxCpuUtilizationPercentage";
const _MD = "MonitorDimension";
const _MDBID = "MemoryDBInstanceDetails";
const _MFC = "MetricForComparison";
const _MI = "MaxImpact";
const _MMUP = "MaxMemoryUtilizationPercentage";
const _MN = "MonitorName";
const _MNOCUUPH = "MinimumNumberOfCapacityUnitsUsedPerHour";
const _MNOCUUPHa = "MaximumNumberOfCapacityUnitsUsedPerHour";
const _MNOIUPH = "MinimumNumberOfInstancesUsedPerHour";
const _MNOIUPHa = "MaximumNumberOfInstancesUsedPerHour";
const _MNUUPH = "MinimumNormalizedUnitsUsedPerHour";
const _MNUUPHa = "MaximumNormalizedUnitsUsedPerHour";
const _MO = "MatchOptions";
const _MOLP = "MetricsOverLookbackPeriod";
const _MR = "MaxResults";
const _MRD = "ModifyRecommendationDetail";
const _MS = "MonitorSpecification";
const _MSUP = "MaxStorageUtilizationPercentage";
const _MSa = "MaxScore";
const _MT = "MonitorType";
const _MV = "MeanValue";
const _MVe = "MetricValue";
const _Me = "Metrics";
const _Mem = "Memory";
const _Met = "Method";
const _Meta = "Metadata";
const _Metr = "Metric";
const _N = "Name";
const _NIBPS = "NetworkInBytesPerSecond";
const _NO = "NumericOperator";
const _NOBPS = "NetworkOutBytesPerSecond";
const _NOR = "NumberOfRules";
const _NP = "NetworkPerformance";
const _NPIPS = "NetworkPacketsInPerSecond";
const _NPOPS = "NetworkPacketsOutPerSecond";
const _NPT = "NextPageToken";
const _NRIS = "NetRISavings";
const _NRU = "NetworkResourceUtilization";
const _NS = "NetSavings";
const _NT = "NodeType";
const _NTe = "NextToken";
const _No = "Not";
const _O = "Or";
const _OC = "OfferingClass";
const _ODC = "OnDemandCost";
const _ODCE = "OnDemandCostEquivalent";
const _ODCORIHU = "OnDemandCostOfRIHoursUsed";
const _ODH = "OnDemandHours";
const _ODHILP = "OnDemandHoursInLookbackPeriod";
const _ODNU = "OnDemandNormalizedUnits";
const _OI = "OfferingId";
const _P = "Parameters";
const _PAF = "ProvideAnomalyFeedback";
const _PAFR = "ProvideAnomalyFeedbackRequest";
const _PAFRr = "ProvideAnomalyFeedbackResponse";
const _PD = "ProductDescription";
const _PDl = "PlatformDifferences";
const _PH = "PurchasedHours";
const _PIL = "PredictionIntervalLevel";
const _PILB = "PredictionIntervalLowerBound";
const _PIUB = "PredictionIntervalUpperBound";
const _PO = "PaymentOption";
const _PS = "ProcessingStatus";
const _PSa = "PageSize";
const _PU = "PurchasedUnits";
const _Pl = "Platform";
const _R = "Rules";
const _RA = "RequestedAt";
const _RAe = "ResourceArn";
const _RAes = "ReservationAggregates";
const _RBT = "ResultsByTime";
const _RBTe = "ResultByTime";
const _RC = "RootCauses";
const _RCD = "ReservedCapacityDetails";
const _RCE = "RequestChangedException";
const _RCG = "ReservationCoverageGroup";
const _RCGe = "ReservationCoverageGroups";
const _RCHILP = "ReservationCoveredHoursInLookbackPeriod";
const _RCI = "RootCauseImpact";
const _RCo = "RootCause";
const _RD = "ResourceDetails";
const _RDD = "RecommendationDetailData";
const _RDHM = "RecommendationDetailHourlyMetrics";
const _RDI = "RecommendationDetailId";
const _RDSID = "RDSInstanceDetails";
const _RDe = "RecommendationDetails";
const _RH = "ReservedHours";
const _RI = "ResourceId";
const _RICFUH = "RICostForUnusedHours";
const _RID = "RedshiftInstanceDetails";
const _RIe = "RecommendationId";
const _RIec = "RecommendationIds";
const _RN = "ResourceName";
const _RNFE = "ResourceNotFoundException";
const _RNOCUTP = "RecommendedNumberOfCapacityUnitsToPurchase";
const _RNOITP = "RecommendedNumberOfInstancesToPurchase";
const _RNU = "ReservedNormalizedUnits";
const _RNUTP = "RecommendedNormalizedUnitsToPurchase";
const _RPR = "ReservationPurchaseRecommendation";
const _RPRD = "ReservationPurchaseRecommendationDetail";
const _RPRDe = "ReservationPurchaseRecommendationDetails";
const _RPRM = "ReservationPurchaseRecommendationMetadata";
const _RPRS = "ReservationPurchaseRecommendationSummary";
const _RPRe = "ReservationPurchaseRecommendations";
const _RR = "RightsizingRecommendations";
const _RRC = "RightsizingRecommendationConfiguration";
const _RRL = "RightsizingRecommendationList";
const _RRM = "RightsizingRecommendationMetadata";
const _RRS = "RightsizingRecommendationSummary";
const _RRi = "RightsizingRecommendation";
const _RS = "ReturnSize";
const _RSMC = "RecurringStandardMonthlyCost";
const _RSe = "RealizedSavings";
const _RSec = "RecommendationSummary";
const _RT = "ResourceTags";
const _RTK = "ResourceTagKeys";
const _RTL = "ResourceTagList";
const _RTe = "ResourceTag";
const _RTec = "RecommendationTarget";
const _RTi = "RightsizingType";
const _RU = "ResourceUtilization";
const _RUG = "ReservationUtilizationGroup";
const _RUGe = "ReservationUtilizationGroups";
const _RV = "RuleVersion";
const _Re = "Region";
const _Rec = "Recommendations";
const _Ru = "Rule";
const _S = "Subscribers";
const _SA = "SubscriptionArn";
const _SAL = "SubscriptionArnList";
const _SB = "SortBy";
const _SCATB = "StartCostAllocationTagBackfill";
const _SCATBR = "StartCostAllocationTagBackfillRequest";
const _SCATBRt = "StartCostAllocationTagBackfillResponse";
const _SCBSP = "SpendCoveredBySavingsPlans";
const _SCC = "SavingsCurrencyCode";
const _SCPA = "StartCommitmentPurchaseAnalysis";
const _SCPAR = "StartCommitmentPurchaseAnalysisRequest";
const _SCPARt = "StartCommitmentPurchaseAnalysisResponse";
const _SCR = "SplitChargeRules";
const _SD = "StartDate";
const _SDo = "SortDefinition";
const _SDor = "SortDefinitions";
const _SFE = "SizeFlexEligible";
const _SN = "SubscriptionName";
const _SO = "SortOrder";
const _SP = "SavingsPercentage";
const _SPA = "SavingsPlanArn";
const _SPAC = "SavingsPlansAmortizedCommitment";
const _SPC = "SavingsPlansCoverages";
const _SPCD = "SavingsPlansCoverageData";
const _SPCHILP = "SavingsPlansCoveredHoursInLookbackPeriod";
const _SPCa = "SavingsPlansCommitment";
const _SPCav = "SavingsPlansCoverage";
const _SPD = "SavingsPlansDetails";
const _SPPAC = "SavingsPlansPurchaseAnalysisConfiguration";
const _SPPAD = "SavingsPlansPurchaseAnalysisDetails";
const _SPPR = "SavingsPlansPurchaseRecommendation";
const _SPPRD = "SavingsPlansPurchaseRecommendationDetails";
const _SPPRDL = "SavingsPlansPurchaseRecommendationDetailList";
const _SPPRDa = "SavingsPlansPurchaseRecommendationDetail";
const _SPPRM = "SavingsPlansPurchaseRecommendationMetadata";
const _SPPRS = "SavingsPlansPurchaseRecommendationSummary";
const _SPS = "SavingsPlansSavings";
const _SPT = "SavingsPlansType";
const _SPTA = "SavingsPlansToAdd";
const _SPTE = "SavingsPlansToExclude";
const _SPU = "SavingsPlansUtilization";
const _SPUA = "SavingsPlansUtilizationAggregates";
const _SPUBT = "SavingsPlansUtilizationsByTime";
const _SPUBTa = "SavingsPlansUtilizationByTime";
const _SPUD = "SavingsPlansUtilizationDetails";
const _SPUDa = "SavingsPlansUtilizationDetail";
const _SPa = "SavingsPlans";
const _SQEE = "ServiceQuotaExceededException";
const _SS = "SearchString";
const _SSPPRG = "StartSavingsPlansPurchaseRecommendationGeneration";
const _SSPPRGR = "StartSavingsPlansPurchaseRecommendationGenerationRequest";
const _SSPPRGRt = "StartSavingsPlansPurchaseRecommendationGenerationResponse";
const _SSe = "ServiceSpecification";
const _ST = "StartTime";
const _SV = "StartValue";
const _Sa = "Savings";
const _Se = "Services";
const _Ser = "Service";
const _Sk = "Sku";
const _So = "Source";
const _St = "Status";
const _Sta = "Start";
const _Sto = "Storage";
const _Su = "Summary";
const _Sub = "Subscriber";
const _T = "Threshold";
const _TAC = "TotalAmortizedCommitment";
const _TAF = "TotalAmortizedFee";
const _TAH = "TotalActualHours";
const _TAS = "TotalActualSpend";
const _TAU = "TotalActualUnits";
const _TC = "TotalCost";
const _TCAU = "TotalCostAndUsage";
const _TCo = "TotalCommitment";
const _TE = "ThresholdExpression";
const _TEMSA = "TotalEstimatedMonthlySavingsAmount";
const _TEMSP = "TotalEstimatedMonthlySavingsPercentage";
const _TES = "TotalExpectedSpend";
const _TI = "TotalImpact";
const _TIF = "TotalImpactFilter";
const _TIL = "TargetInstancesList";
const _TIP = "TotalImpactPercentage";
const _TIY = "TermInYears";
const _TIa = "TargetInstances";
const _TIar = "TargetInstance";
const _TK = "TagKey";
const _TKa = "TagKeys";
const _TMTE = "TooManyTagsException";
const _TP = "TimePeriod";
const _TPRIS = "TotalPotentialRISavings";
const _TR = "TotalRecords";
const _TRC = "TotalRecommendationCount";
const _TRD = "TerminateRecommendationDetail";
const _TRH = "TotalRunningHours";
const _TRHILP = "TotalRunningHoursInLookbackPeriod";
const _TRNU = "TotalRunningNormalizedUnits";
const _TRR = "TagResourceRequest";
const _TRRa = "TagResourceResponse";
const _TRa = "TagResource";
const _TS = "TotalSize";
const _TV = "TagValues";
const _TVL = "TagValuesList";
const _Ta = "Targets";
const _Tag = "Tags";
const _Te = "Tenancy";
const _To = "Total";
const _Ty = "Type";
const _U = "Unit";
const _UAM = "UpdateAnomalyMonitor";
const _UAMR = "UpdateAnomalyMonitorRequest";
const _UAMRp = "UpdateAnomalyMonitorResponse";
const _UAS = "UpdateAnomalySubscription";
const _UASR = "UpdateAnomalySubscriptionRequest";
const _UASRp = "UpdateAnomalySubscriptionResponse";
const _UBT = "UtilizationsByTime";
const _UBTt = "UtilizationByTime";
const _UC = "UpfrontCost";
const _UCATS = "UpdateCostAllocationTagsStatus";
const _UCATSE = "UpdateCostAllocationTagsStatusError";
const _UCATSEp = "UpdateCostAllocationTagsStatusErrors";
const _UCATSR = "UpdateCostAllocationTagsStatusRequest";
const _UCATSRp = "UpdateCostAllocationTagsStatusResponse";
const _UCCD = "UpdateCostCategoryDefinition";
const _UCCDR = "UpdateCostCategoryDefinitionRequest";
const _UCCDRp = "UpdateCostCategoryDefinitionResponse";
const _UCn = "UnusedCommitment";
const _UCs = "UsedCommitment";
const _UH = "UnusedHours";
const _UME = "UnknownMonitorException";
const _UP = "UtilizationPercentage";
const _UPIU = "UtilizationPercentageInUnits";
const _UR = "UntagResource";
const _URR = "UntagResourceRequest";
const _URRn = "UntagResourceResponse";
const _US = "UnrealizedSavings";
const _USE = "UnknownSubscriptionException";
const _UT = "UsageType";
const _UU = "UnusedUnits";
const _UUUE = "UnresolvableUsageUnitException";
const _Ut = "Utilization";
const _V = "Values";
const _Va = "Value";
const _Vc = "Vcpu";
const _c = "client";
const _e = "error";
const _hE = "httpError";
const _s = "smithy.ts.sdk.synthetic.com.amazonaws.costexplorer";
const n0 = "com.amazonaws.costexplorer";
var AnalysisDetails = [
    3,
    n0,
    _AD,
    0,
    [_SPPAD],
    [() => SavingsPlansPurchaseAnalysisDetails],
];
var AnalysisNotFoundException = [
    -3,
    n0,
    _ANFE,
    {
        [_e]: _c,
        [_hE]: 400,
    },
    [_M],
    [0],
];
schema.TypeRegistry.for(n0).registerError(AnalysisNotFoundException, AnalysisNotFoundException$1);
var AnalysisSummary = [
    3,
    n0,
    _AS,
    0,
    [_ECT, _ACT, _AST, _ASn, _EC, _AI, _CPAC],
    [0, 0, 0, 0, 0, 0, () => CommitmentPurchaseAnalysisConfiguration],
];
var Anomaly = [
    3,
    n0,
    _A,
    0,
    [_AIn, _ASD, _AED, _DV, _RC, _ASno, _I, _MA, _F],
    [0, 0, 0, 0, () => RootCauses, () => AnomalyScore, () => Impact, 0, 0],
];
var AnomalyDateInterval = [3, n0, _ADI, 0, [_SD, _ED], [0, 0]];
var AnomalyMonitor = [
    3,
    n0,
    _AM,
    0,
    [_MA, _MN, _CD, _LUD, _LED, _MT, _MD, _MS, _DVC],
    [0, 0, 0, 0, 0, 0, 0, () => Expression, 1],
];
var AnomalyScore = [3, n0, _ASno, 0, [_MSa, _CS], [1, 1]];
var AnomalySubscription = [
    3,
    n0,
    _ASnom,
    0,
    [_SA, _AIc, _MAL, _S, _T, _Fr, _SN, _TE],
    [0, 0, 64 | 0, () => Subscribers, 1, 0, 0, () => Expression],
];
var BackfillLimitExceededException = [
    -3,
    n0,
    _BLEE,
    {
        [_e]: _c,
        [_hE]: 400,
    },
    [_M],
    [0],
];
schema.TypeRegistry.for(n0).registerError(BackfillLimitExceededException, BackfillLimitExceededException$1);
var BillExpirationException = [
    -3,
    n0,
    _BEE,
    {
        [_e]: _c,
    },
    [_M],
    [0],
];
schema.TypeRegistry.for(n0).registerError(BillExpirationException, BillExpirationException$1);
var BillingViewHealthStatusException = [
    -3,
    n0,
    _BVHSE,
    {
        [_e]: _c,
    },
    [_M],
    [0],
];
schema.TypeRegistry.for(n0).registerError(BillingViewHealthStatusException, BillingViewHealthStatusException$1);
var CommitmentPurchaseAnalysisConfiguration = [
    3,
    n0,
    _CPAC,
    0,
    [_SPPAC],
    [() => SavingsPlansPurchaseAnalysisConfiguration],
];
var ComparisonMetricValue = [3, n0, _CMV, 0, [_BTPA, _CTPA, _D, _U], [0, 0, 0, 0]];
var CostAllocationTag = [3, n0, _CAT, 0, [_TK, _Ty, _St, _LUD, _LUDa], [0, 0, 0, 0, 0]];
var CostAllocationTagBackfillRequest = [
    3,
    n0,
    _CATBR,
    0,
    [_BF, _RA, _CA, _BS, _LUA],
    [0, 0, 0, 0, 0],
];
var CostAllocationTagStatusEntry = [3, n0, _CATSE, 0, [_TK, _St], [0, 0]];
var CostAndUsageComparison = [
    3,
    n0,
    _CAUC,
    0,
    [_CAUS, _Me],
    [() => Expression, () => ComparisonMetrics],
];
var CostCategory = [
    3,
    n0,
    _CC,
    0,
    [_CCA, _ES, _EE, _N, _RV, _R, _SCR, _PS, _DVe],
    [
        0,
        0,
        0,
        0,
        0,
        () => CostCategoryRulesList,
        () => CostCategorySplitChargeRulesList,
        () => CostCategoryProcessingStatusList,
        0,
    ],
];
var CostCategoryInheritedValueDimension = [3, n0, _CCIVD, 0, [_DN, _DK], [0, 0]];
var CostCategoryProcessingStatus = [3, n0, _CCPS, 0, [_C, _St], [0, 0]];
var CostCategoryReference = [
    3,
    n0,
    _CCR,
    0,
    [_CCA, _N, _ES, _EE, _NOR, _PS, _V, _DVe],
    [0, 0, 0, 0, 1, () => CostCategoryProcessingStatusList, 64 | 0, 0],
];
var CostCategoryRule = [
    3,
    n0,
    _CCRo,
    0,
    [_Va, _Ru, _IV, _Ty],
    [0, () => Expression, () => CostCategoryInheritedValueDimension, 0],
];
var CostCategorySplitChargeRule = [
    3,
    n0,
    _CCSCR,
    0,
    [_So, _Ta, _Met, _P],
    [0, 64 | 0, 0, () => CostCategorySplitChargeRuleParametersList],
];
var CostCategorySplitChargeRuleParameter = [3, n0, _CCSCRP, 0, [_Ty, _V], [0, 64 | 0]];
var CostCategoryValues = [3, n0, _CCV, 0, [_K, _V, _MO], [0, 64 | 0, 64 | 0]];
var CostComparisonDriver = [
    3,
    n0,
    _CCD,
    0,
    [_CSo, _Me, _CDo],
    [() => Expression, () => ComparisonMetrics, () => CostDrivers],
];
var CostDriver = [3, n0, _CDos, 0, [_Ty, _N, _Me], [0, 0, () => ComparisonMetrics]];
var Coverage = [
    3,
    n0,
    _Co,
    0,
    [_CH, _CNU, _CCo],
    [() => CoverageHours, () => CoverageNormalizedUnits, () => CoverageCost],
];
var CoverageByTime = [
    3,
    n0,
    _CBT,
    0,
    [_TP, _G, _To],
    [() => DateInterval, () => ReservationCoverageGroups, () => Coverage],
];
var CoverageCost = [3, n0, _CCo, 0, [_ODC], [0]];
var CoverageHours = [3, n0, _CH, 0, [_ODH, _RH, _TRH, _CHP], [0, 0, 0, 0]];
var CoverageNormalizedUnits = [3, n0, _CNU, 0, [_ODNU, _RNU, _TRNU, _CNUP], [0, 0, 0, 0]];
var CreateAnomalyMonitorRequest = [
    3,
    n0,
    _CAMR,
    0,
    [_AM, _RT],
    [() => AnomalyMonitor, () => ResourceTagList],
];
var CreateAnomalyMonitorResponse = [3, n0, _CAMRr, 0, [_MA], [0]];
var CreateAnomalySubscriptionRequest = [
    3,
    n0,
    _CASR,
    0,
    [_ASnom, _RT],
    [() => AnomalySubscription, () => ResourceTagList],
];
var CreateAnomalySubscriptionResponse = [3, n0, _CASRr, 0, [_SA], [0]];
var CreateCostCategoryDefinitionRequest = [
    3,
    n0,
    _CCCDR,
    0,
    [_N, _ES, _RV, _R, _DVe, _SCR, _RT],
    [0, 0, 0, () => CostCategoryRulesList, 0, () => CostCategorySplitChargeRulesList, () => ResourceTagList],
];
var CreateCostCategoryDefinitionResponse = [3, n0, _CCCDRr, 0, [_CCA, _ES], [0, 0]];
var CurrentInstance = [
    3,
    n0,
    _CI,
    0,
    [_RI, _IN, _Tag, _RD, _RU, _RCHILP, _SPCHILP, _ODHILP, _TRHILP, _MC, _CCu],
    [0, 0, () => TagValuesList, () => ResourceDetails, () => ResourceUtilization, 0, 0, 0, 0, 0, 0],
];
var DataUnavailableException = [
    -3,
    n0,
    _DUE,
    {
        [_e]: _c,
    },
    [_M],
    [0],
];
schema.TypeRegistry.for(n0).registerError(DataUnavailableException, DataUnavailableException$1);
var DateInterval = [3, n0, _DI, 0, [_Sta, _E], [0, 0]];
var DeleteAnomalyMonitorRequest = [3, n0, _DAMR, 0, [_MA], [0]];
var DeleteAnomalyMonitorResponse = [3, n0, _DAMRe, 0, [], []];
var DeleteAnomalySubscriptionRequest = [3, n0, _DASR, 0, [_SA], [0]];
var DeleteAnomalySubscriptionResponse = [3, n0, _DASRe, 0, [], []];
var DeleteCostCategoryDefinitionRequest = [3, n0, _DCCDR, 0, [_CCA], [0]];
var DeleteCostCategoryDefinitionResponse = [3, n0, _DCCDRe, 0, [_CCA, _EE], [0, 0]];
var DescribeCostCategoryDefinitionRequest = [3, n0, _DCCDRes, 0, [_CCA, _EO], [0, 0]];
var DescribeCostCategoryDefinitionResponse = [
    3,
    n0,
    _DCCDResc,
    0,
    [_CC],
    [() => CostCategory],
];
var DimensionValues = [3, n0, _DVi, 0, [_K, _V, _MO], [0, 64 | 0, 64 | 0]];
var DimensionValuesWithAttributes = [3, n0, _DVWA, 0, [_Va, _At], [0, 128 | 0]];
var DiskResourceUtilization = [
    3,
    n0,
    _DRU,
    0,
    [_DROPS, _DWOPS, _DRBPS, _DWBPS],
    [0, 0, 0, 0],
];
var DynamoDBCapacityDetails = [3, n0, _DDBCD, 0, [_CU, _Re], [0, 0]];
var EBSResourceUtilization = [
    3,
    n0,
    _EBSRU,
    0,
    [_EROPS, _EWOPS, _ERBPS, _EWBPS],
    [0, 0, 0, 0],
];
var EC2InstanceDetails = [
    3,
    n0,
    _ECID,
    0,
    [_Fa, _IT, _Re, _AZ, _Pl, _Te, _CG, _SFE],
    [0, 0, 0, 0, 0, 0, 2, 2],
];
var EC2ResourceDetails = [
    3,
    n0,
    _ECRD,
    0,
    [_HODR, _IT, _Pl, _Re, _Sk, _Mem, _NP, _Sto, _Vc],
    [0, 0, 0, 0, 0, 0, 0, 0, 0],
];
var EC2ResourceUtilization = [
    3,
    n0,
    _ECRU,
    0,
    [_MCUP, _MMUP, _MSUP, _EBSRU, _DRU, _NRU],
    [0, 0, 0, () => EBSResourceUtilization, () => DiskResourceUtilization, () => NetworkResourceUtilization],
];
var EC2Specification = [3, n0, _ECS, 0, [_OC], [0]];
var ElastiCacheInstanceDetails = [
    3,
    n0,
    _ECIDl,
    0,
    [_Fa, _NT, _Re, _PD, _CG, _SFE],
    [0, 0, 0, 0, 2, 2],
];
var ESInstanceDetails = [3, n0, _ESID, 0, [_IC, _IS, _Re, _CG, _SFE], [0, 0, 0, 2, 2]];
var Expression = [
    3,
    n0,
    _Ex,
    0,
    [_O, _An, _No, _Di, _Tag, _CCos],
    [
        () => Expressions,
        () => Expressions,
        () => Expression,
        () => DimensionValues,
        () => TagValues,
        () => CostCategoryValues,
    ],
];
var ForecastResult = [
    3,
    n0,
    _FR,
    0,
    [_TP, _MV, _PILB, _PIUB],
    [() => DateInterval, 0, 0, 0],
];
var GenerationExistsException = [
    -3,
    n0,
    _GEE,
    {
        [_e]: _c,
        [_hE]: 400,
    },
    [_M],
    [0],
];
schema.TypeRegistry.for(n0).registerError(GenerationExistsException, GenerationExistsException$1);
var GenerationSummary = [3, n0, _GS, 0, [_RIe, _GSe, _GST, _GCT, _ECT], [0, 0, 0, 0, 0]];
var GetAnomaliesRequest = [
    3,
    n0,
    _GAR,
    0,
    [_MA, _DI, _F, _TI, _NPT, _MR],
    [0, () => AnomalyDateInterval, 0, () => TotalImpactFilter, 0, 1],
];
var GetAnomaliesResponse = [3, n0, _GARe, 0, [_Ano, _NPT], [() => Anomalies, 0]];
var GetAnomalyMonitorsRequest = [3, n0, _GAMR, 0, [_MAL, _NPT, _MR], [64 | 0, 0, 1]];
var GetAnomalyMonitorsResponse = [
    3,
    n0,
    _GAMRe,
    0,
    [_AMn, _NPT],
    [() => AnomalyMonitors, 0],
];
var GetAnomalySubscriptionsRequest = [
    3,
    n0,
    _GASR,
    0,
    [_SAL, _MA, _NPT, _MR],
    [64 | 0, 0, 0, 1],
];
var GetAnomalySubscriptionsResponse = [
    3,
    n0,
    _GASRe,
    0,
    [_ASnoma, _NPT],
    [() => AnomalySubscriptions, 0],
];
var GetApproximateUsageRecordsRequest = [
    3,
    n0,
    _GAURR,
    0,
    [_Gr, _Se, _ADp],
    [0, 64 | 0, 0],
];
var GetApproximateUsageRecordsResponse = [
    3,
    n0,
    _GAURRe,
    0,
    [_Se, _TR, _LP],
    [128 | 1, 1, () => DateInterval],
];
var GetCommitmentPurchaseAnalysisRequest = [3, n0, _GCPAR, 0, [_AI], [0]];
var GetCommitmentPurchaseAnalysisResponse = [
    3,
    n0,
    _GCPARe,
    0,
    [_ECT, _ACT, _AST, _AI, _ASn, _EC, _AD, _CPAC],
    [0, 0, 0, 0, 0, 0, () => AnalysisDetails, () => CommitmentPurchaseAnalysisConfiguration],
];
var GetCostAndUsageComparisonsRequest = [
    3,
    n0,
    _GCAUCR,
    0,
    [_BVA, _BTP, _CTP, _MFC, _Fi, _GB, _MR, _NPT],
    [0, () => DateInterval, () => DateInterval, 0, () => Expression, () => GroupDefinitions, 1, 0],
];
var GetCostAndUsageComparisonsResponse = [
    3,
    n0,
    _GCAUCRe,
    0,
    [_CAUCo, _TCAU, _NPT],
    [() => CostAndUsageComparisons, () => ComparisonMetrics, 0],
];
var GetCostAndUsageRequest = [
    3,
    n0,
    _GCAUR,
    0,
    [_TP, _Gr, _Fi, _Me, _GB, _BVA, _NPT],
    [() => DateInterval, 0, () => Expression, 64 | 0, () => GroupDefinitions, 0, 0],
];
var GetCostAndUsageResponse = [
    3,
    n0,
    _GCAURe,
    0,
    [_NPT, _GD, _RBT, _DVA],
    [0, () => GroupDefinitions, () => ResultsByTime, () => DimensionValuesWithAttributesList],
];
var GetCostAndUsageWithResourcesRequest = [
    3,
    n0,
    _GCAUWRR,
    0,
    [_TP, _Gr, _Fi, _Me, _GB, _BVA, _NPT],
    [() => DateInterval, 0, () => Expression, 64 | 0, () => GroupDefinitions, 0, 0],
];
var GetCostAndUsageWithResourcesResponse = [
    3,
    n0,
    _GCAUWRRe,
    0,
    [_NPT, _GD, _RBT, _DVA],
    [0, () => GroupDefinitions, () => ResultsByTime, () => DimensionValuesWithAttributesList],
];
var GetCostCategoriesRequest = [
    3,
    n0,
    _GCCR,
    0,
    [_SS, _TP, _CCN, _Fi, _SB, _BVA, _MR, _NPT],
    [0, () => DateInterval, 0, () => Expression, () => SortDefinitions, 0, 1, 0],
];
var GetCostCategoriesResponse = [
    3,
    n0,
    _GCCRe,
    0,
    [_NPT, _CCNo, _CCV, _RS, _TS],
    [0, 64 | 0, 64 | 0, 1, 1],
];
var GetCostComparisonDriversRequest = [
    3,
    n0,
    _GCCDR,
    0,
    [_BVA, _BTP, _CTP, _MFC, _Fi, _GB, _MR, _NPT],
    [0, () => DateInterval, () => DateInterval, 0, () => Expression, () => GroupDefinitions, 1, 0],
];
var GetCostComparisonDriversResponse = [
    3,
    n0,
    _GCCDRe,
    0,
    [_CCDo, _NPT],
    [() => CostComparisonDrivers, 0],
];
var GetCostForecastRequest = [
    3,
    n0,
    _GCFR,
    0,
    [_TP, _Metr, _Gr, _Fi, _BVA, _PIL],
    [() => DateInterval, 0, 0, () => Expression, 0, 1],
];
var GetCostForecastResponse = [
    3,
    n0,
    _GCFRe,
    0,
    [_To, _FRBT],
    [() => MetricValue, () => ForecastResultsByTime],
];
var GetDimensionValuesRequest = [
    3,
    n0,
    _GDVR,
    0,
    [_SS, _TP, _Dim, _Con, _Fi, _SB, _BVA, _MR, _NPT],
    [0, () => DateInterval, 0, 0, () => Expression, () => SortDefinitions, 0, 1, 0],
];
var GetDimensionValuesResponse = [
    3,
    n0,
    _GDVRe,
    0,
    [_DVi, _RS, _TS, _NPT],
    [() => DimensionValuesWithAttributesList, 1, 1, 0],
];
var GetReservationCoverageRequest = [
    3,
    n0,
    _GRCR,
    0,
    [_TP, _GB, _Gr, _Fi, _Me, _NPT, _SB, _MR],
    [() => DateInterval, () => GroupDefinitions, 0, () => Expression, 64 | 0, 0, () => SortDefinition, 1],
];
var GetReservationCoverageResponse = [
    3,
    n0,
    _GRCRe,
    0,
    [_CBTo, _To, _NPT],
    [() => CoveragesByTime, () => Coverage, 0],
];
var GetReservationPurchaseRecommendationRequest = [
    3,
    n0,
    _GRPRR,
    0,
    [_AIc, _Ser, _Fi, _ASc, _LPID, _TIY, _PO, _SSe, _PSa, _NPT],
    [0, 0, () => Expression, 0, 0, 0, 0, () => ServiceSpecification, 1, 0],
];
var GetReservationPurchaseRecommendationResponse = [
    3,
    n0,
    _GRPRRe,
    0,
    [_Meta, _Rec, _NPT],
    [() => ReservationPurchaseRecommendationMetadata, () => ReservationPurchaseRecommendations, 0],
];
var GetReservationUtilizationRequest = [
    3,
    n0,
    _GRUR,
    0,
    [_TP, _GB, _Gr, _Fi, _SB, _NPT, _MR],
    [() => DateInterval, () => GroupDefinitions, 0, () => Expression, () => SortDefinition, 0, 1],
];
var GetReservationUtilizationResponse = [
    3,
    n0,
    _GRURe,
    0,
    [_UBT, _To, _NPT],
    [() => UtilizationsByTime, () => ReservationAggregates, 0],
];
var GetRightsizingRecommendationRequest = [
    3,
    n0,
    _GRRR,
    0,
    [_Fi, _Conf, _Ser, _PSa, _NPT],
    [() => Expression, () => RightsizingRecommendationConfiguration, 0, 1, 0],
];
var GetRightsizingRecommendationResponse = [
    3,
    n0,
    _GRRRe,
    0,
    [_Meta, _Su, _RR, _NPT, _Conf],
    [
        () => RightsizingRecommendationMetadata,
        () => RightsizingRecommendationSummary,
        () => RightsizingRecommendationList,
        0,
        () => RightsizingRecommendationConfiguration,
    ],
];
var GetSavingsPlanPurchaseRecommendationDetailsRequest = [
    3,
    n0,
    _GSPPRDR,
    0,
    [_RDI],
    [0],
];
var GetSavingsPlanPurchaseRecommendationDetailsResponse = [
    3,
    n0,
    _GSPPRDRe,
    0,
    [_RDI, _RDD],
    [0, () => RecommendationDetailData],
];
var GetSavingsPlansCoverageRequest = [
    3,
    n0,
    _GSPCR,
    0,
    [_TP, _GB, _Gr, _Fi, _Me, _NTe, _MR, _SB],
    [() => DateInterval, () => GroupDefinitions, 0, () => Expression, 64 | 0, 0, 1, () => SortDefinition],
];
var GetSavingsPlansCoverageResponse = [
    3,
    n0,
    _GSPCRe,
    0,
    [_SPC, _NTe],
    [() => SavingsPlansCoverages, 0],
];
var GetSavingsPlansPurchaseRecommendationRequest = [
    3,
    n0,
    _GSPPRR,
    0,
    [_SPT, _TIY, _PO, _ASc, _NPT, _PSa, _LPID, _Fi],
    [0, 0, 0, 0, 0, 1, 0, () => Expression],
];
var GetSavingsPlansPurchaseRecommendationResponse = [
    3,
    n0,
    _GSPPRRe,
    0,
    [_Meta, _SPPR, _NPT],
    [() => SavingsPlansPurchaseRecommendationMetadata, () => SavingsPlansPurchaseRecommendation, 0],
];
var GetSavingsPlansUtilizationDetailsRequest = [
    3,
    n0,
    _GSPUDR,
    0,
    [_TP, _Fi, _DT, _NTe, _MR, _SB],
    [() => DateInterval, () => Expression, 64 | 0, 0, 1, () => SortDefinition],
];
var GetSavingsPlansUtilizationDetailsResponse = [
    3,
    n0,
    _GSPUDRe,
    0,
    [_SPUD, _To, _TP, _NTe],
    [() => SavingsPlansUtilizationDetails, () => SavingsPlansUtilizationAggregates, () => DateInterval, 0],
];
var GetSavingsPlansUtilizationRequest = [
    3,
    n0,
    _GSPUR,
    0,
    [_TP, _Gr, _Fi, _SB],
    [() => DateInterval, 0, () => Expression, () => SortDefinition],
];
var GetSavingsPlansUtilizationResponse = [
    3,
    n0,
    _GSPURe,
    0,
    [_SPUBT, _To],
    [() => SavingsPlansUtilizationsByTime, () => SavingsPlansUtilizationAggregates],
];
var GetTagsRequest = [
    3,
    n0,
    _GTR,
    0,
    [_SS, _TP, _TK, _Fi, _SB, _BVA, _MR, _NPT],
    [0, () => DateInterval, 0, () => Expression, () => SortDefinitions, 0, 1, 0],
];
var GetTagsResponse = [3, n0, _GTRe, 0, [_NPT, _Tag, _RS, _TS], [0, 64 | 0, 1, 1]];
var GetUsageForecastRequest = [
    3,
    n0,
    _GUFR,
    0,
    [_TP, _Metr, _Gr, _Fi, _BVA, _PIL],
    [() => DateInterval, 0, 0, () => Expression, 0, 1],
];
var GetUsageForecastResponse = [
    3,
    n0,
    _GUFRe,
    0,
    [_To, _FRBT],
    [() => MetricValue, () => ForecastResultsByTime],
];
var Group = [3, n0, _Gro, 0, [_Ke, _Me], [64 | 0, () => Metrics]];
var GroupDefinition = [3, n0, _GDr, 0, [_Ty, _K], [0, 0]];
var Impact = [3, n0, _I, 0, [_MI, _TI, _TAS, _TES, _TIP], [1, 1, 1, 1, 1]];
var InstanceDetails = [
    3,
    n0,
    _ID,
    0,
    [_ECID, _RDSID, _RID, _ECIDl, _ESID, _MDBID],
    [
        () => EC2InstanceDetails,
        () => RDSInstanceDetails,
        () => RedshiftInstanceDetails,
        () => ElastiCacheInstanceDetails,
        () => ESInstanceDetails,
        () => MemoryDBInstanceDetails,
    ],
];
var InvalidNextTokenException = [
    -3,
    n0,
    _INTE,
    {
        [_e]: _c,
    },
    [_M],
    [0],
];
schema.TypeRegistry.for(n0).registerError(InvalidNextTokenException, InvalidNextTokenException$1);
var LimitExceededException = [
    -3,
    n0,
    _LEE,
    {
        [_e]: _c,
    },
    [_M],
    [0],
];
schema.TypeRegistry.for(n0).registerError(LimitExceededException, LimitExceededException$1);
var ListCommitmentPurchaseAnalysesRequest = [
    3,
    n0,
    _LCPAR,
    0,
    [_ASn, _NPT, _PSa, _AIna],
    [0, 0, 1, 64 | 0],
];
var ListCommitmentPurchaseAnalysesResponse = [
    3,
    n0,
    _LCPARi,
    0,
    [_ASL, _NPT],
    [() => AnalysisSummaryList, 0],
];
var ListCostAllocationTagBackfillHistoryRequest = [
    3,
    n0,
    _LCATBHR,
    0,
    [_NTe, _MR],
    [0, 1],
];
var ListCostAllocationTagBackfillHistoryResponse = [
    3,
    n0,
    _LCATBHRi,
    0,
    [_BR, _NTe],
    [() => CostAllocationTagBackfillRequestList, 0],
];
var ListCostAllocationTagsRequest = [
    3,
    n0,
    _LCATR,
    0,
    [_St, _TKa, _Ty, _NTe, _MR],
    [0, 64 | 0, 0, 0, 1],
];
var ListCostAllocationTagsResponse = [
    3,
    n0,
    _LCATRi,
    0,
    [_CATo, _NTe],
    [() => CostAllocationTagList, 0],
];
var ListCostCategoryDefinitionsRequest = [3, n0, _LCCDR, 0, [_EO, _NTe, _MR], [0, 0, 1]];
var ListCostCategoryDefinitionsResponse = [
    3,
    n0,
    _LCCDRi,
    0,
    [_CCRos, _NTe],
    [() => CostCategoryReferencesList, 0],
];
var ListSavingsPlansPurchaseRecommendationGenerationRequest = [
    3,
    n0,
    _LSPPRGR,
    0,
    [_GSe, _RIec, _PSa, _NPT],
    [0, 64 | 0, 1, 0],
];
var ListSavingsPlansPurchaseRecommendationGenerationResponse = [
    3,
    n0,
    _LSPPRGRi,
    0,
    [_GSL, _NPT],
    [() => GenerationSummaryList, 0],
];
var ListTagsForResourceRequest = [3, n0, _LTFRR, 0, [_RAe], [0]];
var ListTagsForResourceResponse = [3, n0, _LTFRRi, 0, [_RT], [() => ResourceTagList]];
var MemoryDBInstanceDetails = [
    3,
    n0,
    _MDBID,
    0,
    [_Fa, _NT, _Re, _CG, _SFE],
    [0, 0, 0, 2, 2],
];
var MetricValue = [3, n0, _MVe, 0, [_Am, _U], [0, 0]];
var ModifyRecommendationDetail = [3, n0, _MRD, 0, [_TIa], [() => TargetInstancesList]];
var NetworkResourceUtilization = [
    3,
    n0,
    _NRU,
    0,
    [_NIBPS, _NOBPS, _NPIPS, _NPOPS],
    [0, 0, 0, 0],
];
var ProvideAnomalyFeedbackRequest = [3, n0, _PAFR, 0, [_AIn, _F], [0, 0]];
var ProvideAnomalyFeedbackResponse = [3, n0, _PAFRr, 0, [_AIn], [0]];
var RDSInstanceDetails = [
    3,
    n0,
    _RDSID,
    0,
    [_Fa, _IT, _Re, _DE, _DEa, _DO, _LM, _CG, _SFE],
    [0, 0, 0, 0, 0, 0, 0, 2, 2],
];
var RecommendationDetailData = [
    3,
    n0,
    _RDD,
    0,
    [
        _ASc,
        _LPID,
        _SPT,
        _TIY,
        _PO,
        _AIc,
        _CCu,
        _IF,
        _Re,
        _OI,
        _GT,
        _LUT,
        _CAHODS,
        _CMHODS,
        _CMHODSu,
        _EAU,
        _EMSA,
        _EODC,
        _EODCWCC,
        _EROI,
        _ESPC,
        _ESA,
        _ESP,
        _EHC,
        _HCTP,
        _UC,
        _CAC,
        _EAC,
        _MOLP,
    ],
    [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, () => MetricsOverLookbackPeriod],
];
var RecommendationDetailHourlyMetrics = [
    3,
    n0,
    _RDHM,
    0,
    [_ST, _EODC, _CCur, _ECs, _ENCU],
    [0, 0, 0, 0, 0],
];
var RedshiftInstanceDetails = [
    3,
    n0,
    _RID,
    0,
    [_Fa, _NT, _Re, _CG, _SFE],
    [0, 0, 0, 2, 2],
];
var RequestChangedException = [
    -3,
    n0,
    _RCE,
    {
        [_e]: _c,
    },
    [_M],
    [0],
];
schema.TypeRegistry.for(n0).registerError(RequestChangedException, RequestChangedException$1);
var ReservationAggregates = [
    3,
    n0,
    _RAes,
    0,
    [_UP, _UPIU, _PH, _PU, _TAH, _TAU, _UH, _UU, _ODCORIHU, _NRIS, _TPRIS, _AUF, _ARF, _TAF, _RICFUH, _RSe, _US],
    [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
];
var ReservationCoverageGroup = [3, n0, _RCG, 0, [_At, _Co], [128 | 0, () => Coverage]];
var ReservationPurchaseRecommendation = [
    3,
    n0,
    _RPR,
    0,
    [_ASc, _LPID, _TIY, _PO, _SSe, _RDe, _RSec],
    [
        0,
        0,
        0,
        0,
        () => ServiceSpecification,
        () => ReservationPurchaseRecommendationDetails,
        () => ReservationPurchaseRecommendationSummary,
    ],
];
var ReservationPurchaseRecommendationDetail = [
    3,
    n0,
    _RPRD,
    0,
    [
        _AIc,
        _ID,
        _RNOITP,
        _RNUTP,
        _MNOIUPH,
        _MNUUPH,
        _MNOIUPHa,
        _MNUUPHa,
        _ANOIUPH,
        _ANUUPH,
        _AU,
        _EBEIM,
        _CCu,
        _EMSA,
        _EMSP,
        _EMODC,
        _ERCFLP,
        _UC,
        _RSMC,
        _RCD,
        _RNOCUTP,
        _MNOCUUPH,
        _MNOCUUPHa,
        _ANOCUUPH,
    ],
    [
        0,
        () => InstanceDetails,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        () => ReservedCapacityDetails,
        0,
        0,
        0,
        0,
    ],
];
var ReservationPurchaseRecommendationMetadata = [
    3,
    n0,
    _RPRM,
    0,
    [_RIe, _GT, _AMd],
    [0, 0, 0],
];
var ReservationPurchaseRecommendationSummary = [
    3,
    n0,
    _RPRS,
    0,
    [_TEMSA, _TEMSP, _CCu],
    [0, 0, 0],
];
var ReservationUtilizationGroup = [
    3,
    n0,
    _RUG,
    0,
    [_K, _Va, _At, _Ut],
    [0, 0, 128 | 0, () => ReservationAggregates],
];
var ReservedCapacityDetails = [3, n0, _RCD, 0, [_DDBCD], [() => DynamoDBCapacityDetails]];
var ResourceDetails = [3, n0, _RD, 0, [_ECRD], [() => EC2ResourceDetails]];
var ResourceNotFoundException = [
    -3,
    n0,
    _RNFE,
    {
        [_e]: _c,
        [_hE]: 404,
    },
    [_M, _RN],
    [0, 0],
];
schema.TypeRegistry.for(n0).registerError(ResourceNotFoundException, ResourceNotFoundException$1);
var ResourceTag = [3, n0, _RTe, 0, [_K, _Va], [0, 0]];
var ResourceUtilization = [3, n0, _RU, 0, [_ECRU], [() => EC2ResourceUtilization]];
var ResultByTime = [
    3,
    n0,
    _RBTe,
    0,
    [_TP, _To, _G, _Es],
    [() => DateInterval, () => Metrics, () => Groups, 2],
];
var RightsizingRecommendation = [
    3,
    n0,
    _RRi,
    0,
    [_AIc, _CI, _RTi, _MRD, _TRD, _FRC],
    [0, () => CurrentInstance, 0, () => ModifyRecommendationDetail, () => TerminateRecommendationDetail, 64 | 0],
];
var RightsizingRecommendationConfiguration = [3, n0, _RRC, 0, [_RTec, _BC], [0, 2]];
var RightsizingRecommendationMetadata = [
    3,
    n0,
    _RRM,
    0,
    [_RIe, _GT, _LPID, _AMd],
    [0, 0, 0, 0],
];
var RightsizingRecommendationSummary = [
    3,
    n0,
    _RRS,
    0,
    [_TRC, _ETMSA, _SCC, _SP],
    [0, 0, 0, 0],
];
var RootCause = [
    3,
    n0,
    _RCo,
    0,
    [_Ser, _Re, _LA, _LAN, _UT, _I],
    [0, 0, 0, 0, 0, () => RootCauseImpact],
];
var RootCauseImpact = [3, n0, _RCI, 0, [_Cont], [1]];
var SavingsPlans = [
    3,
    n0,
    _SPa,
    0,
    [_PO, _SPT, _Re, _IF, _TIY, _SPCa, _OI],
    [0, 0, 0, 0, 0, 1, 0],
];
var SavingsPlansAmortizedCommitment = [3, n0, _SPAC, 0, [_ARC, _AUC, _TAC], [0, 0, 0]];
var SavingsPlansCoverage = [
    3,
    n0,
    _SPCav,
    0,
    [_At, _Co, _TP],
    [128 | 0, () => SavingsPlansCoverageData, () => DateInterval],
];
var SavingsPlansCoverageData = [3, n0, _SPCD, 0, [_SCBSP, _ODC, _TC, _CP], [0, 0, 0, 0]];
var SavingsPlansDetails = [3, n0, _SPD, 0, [_Re, _IF, _OI], [0, 0, 0]];
var SavingsPlansPurchaseAnalysisConfiguration = [
    3,
    n0,
    _SPPAC,
    0,
    [_ASc, _AIc, _AT, _SPTA, _SPTE, _LBTP],
    [0, 0, 0, () => SavingsPlansToAdd, 64 | 0, () => DateInterval],
];
var SavingsPlansPurchaseAnalysisDetails = [
    3,
    n0,
    _SPPAD,
    0,
    [
        _CCu,
        _LPIH,
        _CAC,
        _CAHODS,
        _CMHODS,
        _CMHODSu,
        _CODS,
        _EHC,
        _HCTP,
        _EAC,
        _EAU,
        _EMSA,
        _EODC,
        _EODCWCC,
        _EROI,
        _ESA,
        _ESP,
        _ECC,
        _LUT,
        _UC,
        _AMd,
        _MOLP,
    ],
    [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, () => MetricsOverLookbackPeriod],
];
var SavingsPlansPurchaseRecommendation = [
    3,
    n0,
    _SPPR,
    0,
    [_ASc, _SPT, _TIY, _PO, _LPID, _SPPRD, _SPPRS],
    [0, 0, 0, 0, 0, () => SavingsPlansPurchaseRecommendationDetailList, () => SavingsPlansPurchaseRecommendationSummary],
];
var SavingsPlansPurchaseRecommendationDetail = [
    3,
    n0,
    _SPPRDa,
    0,
    [
        _SPD,
        _AIc,
        _UC,
        _EROI,
        _CCu,
        _ESPC,
        _EODC,
        _EODCWCC,
        _ESA,
        _ESP,
        _HCTP,
        _EAU,
        _EMSA,
        _CMHODSu,
        _CMHODS,
        _CAHODS,
        _RDI,
    ],
    [() => SavingsPlansDetails, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
];
var SavingsPlansPurchaseRecommendationMetadata = [
    3,
    n0,
    _SPPRM,
    0,
    [_RIe, _GT, _AMd],
    [0, 0, 0],
];
var SavingsPlansPurchaseRecommendationSummary = [
    3,
    n0,
    _SPPRS,
    0,
    [_EROI, _CCu, _ETC, _CODS, _ESA, _TRC, _DCTP, _HCTP, _ESP, _EMSA, _EODCWCC],
    [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
];
var SavingsPlansSavings = [3, n0, _SPS, 0, [_NS, _ODCE], [0, 0]];
var SavingsPlansUtilization = [3, n0, _SPU, 0, [_TCo, _UCs, _UCn, _UP], [0, 0, 0, 0]];
var SavingsPlansUtilizationAggregates = [
    3,
    n0,
    _SPUA,
    0,
    [_Ut, _Sa, _AC],
    [() => SavingsPlansUtilization, () => SavingsPlansSavings, () => SavingsPlansAmortizedCommitment],
];
var SavingsPlansUtilizationByTime = [
    3,
    n0,
    _SPUBTa,
    0,
    [_TP, _Ut, _Sa, _AC],
    [() => DateInterval, () => SavingsPlansUtilization, () => SavingsPlansSavings, () => SavingsPlansAmortizedCommitment],
];
var SavingsPlansUtilizationDetail = [
    3,
    n0,
    _SPUDa,
    0,
    [_SPA, _At, _Ut, _Sa, _AC],
    [0, 128 | 0, () => SavingsPlansUtilization, () => SavingsPlansSavings, () => SavingsPlansAmortizedCommitment],
];
var ServiceQuotaExceededException = [
    -3,
    n0,
    _SQEE,
    {
        [_e]: _c,
        [_hE]: 402,
    },
    [_M],
    [0],
];
schema.TypeRegistry.for(n0).registerError(ServiceQuotaExceededException, ServiceQuotaExceededException$1);
var ServiceSpecification = [3, n0, _SSe, 0, [_ECS], [() => EC2Specification]];
var SortDefinition = [3, n0, _SDo, 0, [_K, _SO], [0, 0]];
var StartCommitmentPurchaseAnalysisRequest = [
    3,
    n0,
    _SCPAR,
    0,
    [_CPAC],
    [() => CommitmentPurchaseAnalysisConfiguration],
];
var StartCommitmentPurchaseAnalysisResponse = [
    3,
    n0,
    _SCPARt,
    0,
    [_AI, _AST, _ECT],
    [0, 0, 0],
];
var StartCostAllocationTagBackfillRequest = [3, n0, _SCATBR, 0, [_BF], [0]];
var StartCostAllocationTagBackfillResponse = [
    3,
    n0,
    _SCATBRt,
    0,
    [_BRa],
    [() => CostAllocationTagBackfillRequest],
];
var StartSavingsPlansPurchaseRecommendationGenerationRequest = [
    3,
    n0,
    _SSPPRGR,
    0,
    [],
    [],
];
var StartSavingsPlansPurchaseRecommendationGenerationResponse = [
    3,
    n0,
    _SSPPRGRt,
    0,
    [_RIe, _GST, _ECT],
    [0, 0, 0],
];
var Subscriber = [3, n0, _Sub, 0, [_Ad, _Ty, _St], [0, 0, 0]];
var TagResourceRequest = [3, n0, _TRR, 0, [_RAe, _RT], [0, () => ResourceTagList]];
var TagResourceResponse = [3, n0, _TRRa, 0, [], []];
var TagValues = [3, n0, _TV, 0, [_K, _V, _MO], [0, 64 | 0, 64 | 0]];
var TargetInstance = [
    3,
    n0,
    _TIar,
    0,
    [_EMC, _EMS, _CCu, _DTI, _RD, _ERU, _PDl],
    [0, 0, 0, 2, () => ResourceDetails, () => ResourceUtilization, 64 | 0],
];
var TerminateRecommendationDetail = [3, n0, _TRD, 0, [_EMS, _CCu], [0, 0]];
var TooManyTagsException = [
    -3,
    n0,
    _TMTE,
    {
        [_e]: _c,
        [_hE]: 400,
    },
    [_M, _RN],
    [0, 0],
];
schema.TypeRegistry.for(n0).registerError(TooManyTagsException, TooManyTagsException$1);
var TotalImpactFilter = [3, n0, _TIF, 0, [_NO, _SV, _EV], [0, 1, 1]];
var UnknownMonitorException = [
    -3,
    n0,
    _UME,
    {
        [_e]: _c,
        [_hE]: 404,
    },
    [_M],
    [0],
];
schema.TypeRegistry.for(n0).registerError(UnknownMonitorException, UnknownMonitorException$1);
var UnknownSubscriptionException = [
    -3,
    n0,
    _USE,
    {
        [_e]: _c,
        [_hE]: 404,
    },
    [_M],
    [0],
];
schema.TypeRegistry.for(n0).registerError(UnknownSubscriptionException, UnknownSubscriptionException$1);
var UnresolvableUsageUnitException = [
    -3,
    n0,
    _UUUE,
    {
        [_e]: _c,
    },
    [_M],
    [0],
];
schema.TypeRegistry.for(n0).registerError(UnresolvableUsageUnitException, UnresolvableUsageUnitException$1);
var UntagResourceRequest = [3, n0, _URR, 0, [_RAe, _RTK], [0, 64 | 0]];
var UntagResourceResponse = [3, n0, _URRn, 0, [], []];
var UpdateAnomalyMonitorRequest = [3, n0, _UAMR, 0, [_MA, _MN], [0, 0]];
var UpdateAnomalyMonitorResponse = [3, n0, _UAMRp, 0, [_MA], [0]];
var UpdateAnomalySubscriptionRequest = [
    3,
    n0,
    _UASR,
    0,
    [_SA, _T, _Fr, _MAL, _S, _SN, _TE],
    [0, 1, 0, 64 | 0, () => Subscribers, 0, () => Expression],
];
var UpdateAnomalySubscriptionResponse = [3, n0, _UASRp, 0, [_SA], [0]];
var UpdateCostAllocationTagsStatusError = [3, n0, _UCATSE, 0, [_TK, _Cod, _M], [0, 0, 0]];
var UpdateCostAllocationTagsStatusRequest = [
    3,
    n0,
    _UCATSR,
    0,
    [_CATS],
    [() => CostAllocationTagStatusList],
];
var UpdateCostAllocationTagsStatusResponse = [
    3,
    n0,
    _UCATSRp,
    0,
    [_Er],
    [() => UpdateCostAllocationTagsStatusErrors],
];
var UpdateCostCategoryDefinitionRequest = [
    3,
    n0,
    _UCCDR,
    0,
    [_CCA, _ES, _RV, _R, _DVe, _SCR],
    [0, 0, 0, () => CostCategoryRulesList, 0, () => CostCategorySplitChargeRulesList],
];
var UpdateCostCategoryDefinitionResponse = [3, n0, _UCCDRp, 0, [_CCA, _ES], [0, 0]];
var UtilizationByTime = [
    3,
    n0,
    _UBTt,
    0,
    [_TP, _G, _To],
    [() => DateInterval, () => ReservationUtilizationGroups, () => ReservationAggregates],
];
var CostExplorerServiceException = [-3, _s, "CostExplorerServiceException", 0, [], []];
schema.TypeRegistry.for(_s).registerError(CostExplorerServiceException, CostExplorerServiceException$1);
var AnalysisSummaryList = [1, n0, _ASL, 0, () => AnalysisSummary];
var Anomalies = [1, n0, _Ano, 0, () => Anomaly];
var AnomalyMonitors = [1, n0, _AMn, 0, () => AnomalyMonitor];
var AnomalySubscriptions = [1, n0, _ASnoma, 0, () => AnomalySubscription];
var CostAllocationTagBackfillRequestList = [
    1,
    n0,
    _CATBRL,
    0,
    () => CostAllocationTagBackfillRequest,
];
var CostAllocationTagList = [1, n0, _CATL, 0, () => CostAllocationTag];
var CostAllocationTagStatusList = [1, n0, _CATSL, 0, () => CostAllocationTagStatusEntry];
var CostAndUsageComparisons = [1, n0, _CAUCo, 0, () => CostAndUsageComparison];
var CostCategoryProcessingStatusList = [1, n0, _CCPSL, 0, () => CostCategoryProcessingStatus];
var CostCategoryReferencesList = [1, n0, _CCRL, 0, () => CostCategoryReference];
var CostCategoryRulesList = [1, n0, _CCRLo, 0, () => CostCategoryRule];
var CostCategorySplitChargeRuleParametersList = [
    1,
    n0,
    _CCSCRPL,
    0,
    () => CostCategorySplitChargeRuleParameter,
];
var CostCategorySplitChargeRulesList = [1, n0, _CCSCRL, 0, () => CostCategorySplitChargeRule];
var CostComparisonDrivers = [1, n0, _CCDo, 0, () => CostComparisonDriver];
var CostDrivers = [1, n0, _CDo, 0, () => CostDriver];
var CoveragesByTime = [1, n0, _CBTo, 0, () => CoverageByTime];
var DimensionValuesWithAttributesList = [
    1,
    n0,
    _DVWAL,
    0,
    () => DimensionValuesWithAttributes,
];
var Expressions = [1, n0, _Exp, 0, () => Expression];
var ForecastResultsByTime = [1, n0, _FRBT, 0, () => ForecastResult];
var GenerationSummaryList = [1, n0, _GSL, 0, () => GenerationSummary];
var GroupDefinitions = [1, n0, _GD, 0, () => GroupDefinition];
var Groups = [1, n0, _G, 0, () => Group];
var MetricsOverLookbackPeriod = [1, n0, _MOLP, 0, () => RecommendationDetailHourlyMetrics];
var ReservationCoverageGroups = [1, n0, _RCGe, 0, () => ReservationCoverageGroup];
var ReservationPurchaseRecommendationDetails = [
    1,
    n0,
    _RPRDe,
    0,
    () => ReservationPurchaseRecommendationDetail,
];
var ReservationPurchaseRecommendations = [
    1,
    n0,
    _RPRe,
    0,
    () => ReservationPurchaseRecommendation,
];
var ReservationUtilizationGroups = [1, n0, _RUGe, 0, () => ReservationUtilizationGroup];
var ResourceTagList = [1, n0, _RTL, 0, () => ResourceTag];
var ResultsByTime = [1, n0, _RBT, 0, () => ResultByTime];
var RightsizingRecommendationList = [1, n0, _RRL, 0, () => RightsizingRecommendation];
var RootCauses = [1, n0, _RC, 0, () => RootCause];
var SavingsPlansCoverages = [1, n0, _SPC, 0, () => SavingsPlansCoverage];
var SavingsPlansPurchaseRecommendationDetailList = [
    1,
    n0,
    _SPPRDL,
    0,
    () => SavingsPlansPurchaseRecommendationDetail,
];
var SavingsPlansToAdd = [1, n0, _SPTA, 0, () => SavingsPlans];
var SavingsPlansUtilizationDetails = [1, n0, _SPUD, 0, () => SavingsPlansUtilizationDetail];
var SavingsPlansUtilizationsByTime = [1, n0, _SPUBT, 0, () => SavingsPlansUtilizationByTime];
var SortDefinitions = [1, n0, _SDor, 0, () => SortDefinition];
var Subscribers = [1, n0, _S, 0, () => Subscriber];
var TagValuesList = [1, n0, _TVL, 0, () => TagValues];
var TargetInstancesList = [1, n0, _TIL, 0, () => TargetInstance];
var UpdateCostAllocationTagsStatusErrors = [
    1,
    n0,
    _UCATSEp,
    0,
    () => UpdateCostAllocationTagsStatusError,
];
var UtilizationsByTime = [1, n0, _UBT, 0, () => UtilizationByTime];
var ComparisonMetrics = [2, n0, _CM, 0, 0, () => ComparisonMetricValue];
var Metrics = [2, n0, _Me, 0, 0, () => MetricValue];
var CreateAnomalyMonitor = [
    9,
    n0,
    _CAM,
    0,
    () => CreateAnomalyMonitorRequest,
    () => CreateAnomalyMonitorResponse,
];
var CreateAnomalySubscription = [
    9,
    n0,
    _CAS,
    0,
    () => CreateAnomalySubscriptionRequest,
    () => CreateAnomalySubscriptionResponse,
];
var CreateCostCategoryDefinition = [
    9,
    n0,
    _CCCD,
    0,
    () => CreateCostCategoryDefinitionRequest,
    () => CreateCostCategoryDefinitionResponse,
];
var DeleteAnomalyMonitor = [
    9,
    n0,
    _DAM,
    0,
    () => DeleteAnomalyMonitorRequest,
    () => DeleteAnomalyMonitorResponse,
];
var DeleteAnomalySubscription = [
    9,
    n0,
    _DAS,
    0,
    () => DeleteAnomalySubscriptionRequest,
    () => DeleteAnomalySubscriptionResponse,
];
var DeleteCostCategoryDefinition = [
    9,
    n0,
    _DCCD,
    0,
    () => DeleteCostCategoryDefinitionRequest,
    () => DeleteCostCategoryDefinitionResponse,
];
var DescribeCostCategoryDefinition = [
    9,
    n0,
    _DCCDe,
    0,
    () => DescribeCostCategoryDefinitionRequest,
    () => DescribeCostCategoryDefinitionResponse,
];
var GetAnomalies = [9, n0, _GA, 0, () => GetAnomaliesRequest, () => GetAnomaliesResponse];
var GetAnomalyMonitors = [
    9,
    n0,
    _GAM,
    0,
    () => GetAnomalyMonitorsRequest,
    () => GetAnomalyMonitorsResponse,
];
var GetAnomalySubscriptions = [
    9,
    n0,
    _GAS,
    0,
    () => GetAnomalySubscriptionsRequest,
    () => GetAnomalySubscriptionsResponse,
];
var GetApproximateUsageRecords = [
    9,
    n0,
    _GAUR,
    0,
    () => GetApproximateUsageRecordsRequest,
    () => GetApproximateUsageRecordsResponse,
];
var GetCommitmentPurchaseAnalysis = [
    9,
    n0,
    _GCPA,
    0,
    () => GetCommitmentPurchaseAnalysisRequest,
    () => GetCommitmentPurchaseAnalysisResponse,
];
var GetCostAndUsage = [
    9,
    n0,
    _GCAU,
    0,
    () => GetCostAndUsageRequest,
    () => GetCostAndUsageResponse,
];
var GetCostAndUsageComparisons = [
    9,
    n0,
    _GCAUC,
    0,
    () => GetCostAndUsageComparisonsRequest,
    () => GetCostAndUsageComparisonsResponse,
];
var GetCostAndUsageWithResources = [
    9,
    n0,
    _GCAUWR,
    0,
    () => GetCostAndUsageWithResourcesRequest,
    () => GetCostAndUsageWithResourcesResponse,
];
var GetCostCategories = [
    9,
    n0,
    _GCC,
    0,
    () => GetCostCategoriesRequest,
    () => GetCostCategoriesResponse,
];
var GetCostComparisonDrivers = [
    9,
    n0,
    _GCCD,
    0,
    () => GetCostComparisonDriversRequest,
    () => GetCostComparisonDriversResponse,
];
var GetCostForecast = [
    9,
    n0,
    _GCF,
    0,
    () => GetCostForecastRequest,
    () => GetCostForecastResponse,
];
var GetDimensionValues = [
    9,
    n0,
    _GDV,
    0,
    () => GetDimensionValuesRequest,
    () => GetDimensionValuesResponse,
];
var GetReservationCoverage = [
    9,
    n0,
    _GRC,
    0,
    () => GetReservationCoverageRequest,
    () => GetReservationCoverageResponse,
];
var GetReservationPurchaseRecommendation = [
    9,
    n0,
    _GRPR,
    0,
    () => GetReservationPurchaseRecommendationRequest,
    () => GetReservationPurchaseRecommendationResponse,
];
var GetReservationUtilization = [
    9,
    n0,
    _GRU,
    0,
    () => GetReservationUtilizationRequest,
    () => GetReservationUtilizationResponse,
];
var GetRightsizingRecommendation = [
    9,
    n0,
    _GRR,
    0,
    () => GetRightsizingRecommendationRequest,
    () => GetRightsizingRecommendationResponse,
];
var GetSavingsPlanPurchaseRecommendationDetails = [
    9,
    n0,
    _GSPPRD,
    0,
    () => GetSavingsPlanPurchaseRecommendationDetailsRequest,
    () => GetSavingsPlanPurchaseRecommendationDetailsResponse,
];
var GetSavingsPlansCoverage = [
    9,
    n0,
    _GSPC,
    0,
    () => GetSavingsPlansCoverageRequest,
    () => GetSavingsPlansCoverageResponse,
];
var GetSavingsPlansPurchaseRecommendation = [
    9,
    n0,
    _GSPPR,
    0,
    () => GetSavingsPlansPurchaseRecommendationRequest,
    () => GetSavingsPlansPurchaseRecommendationResponse,
];
var GetSavingsPlansUtilization = [
    9,
    n0,
    _GSPU,
    0,
    () => GetSavingsPlansUtilizationRequest,
    () => GetSavingsPlansUtilizationResponse,
];
var GetSavingsPlansUtilizationDetails = [
    9,
    n0,
    _GSPUD,
    0,
    () => GetSavingsPlansUtilizationDetailsRequest,
    () => GetSavingsPlansUtilizationDetailsResponse,
];
var GetTags = [9, n0, _GTe, 0, () => GetTagsRequest, () => GetTagsResponse];
var GetUsageForecast = [
    9,
    n0,
    _GUF,
    0,
    () => GetUsageForecastRequest,
    () => GetUsageForecastResponse,
];
var ListCommitmentPurchaseAnalyses = [
    9,
    n0,
    _LCPA,
    0,
    () => ListCommitmentPurchaseAnalysesRequest,
    () => ListCommitmentPurchaseAnalysesResponse,
];
var ListCostAllocationTagBackfillHistory = [
    9,
    n0,
    _LCATBH,
    0,
    () => ListCostAllocationTagBackfillHistoryRequest,
    () => ListCostAllocationTagBackfillHistoryResponse,
];
var ListCostAllocationTags = [
    9,
    n0,
    _LCAT,
    0,
    () => ListCostAllocationTagsRequest,
    () => ListCostAllocationTagsResponse,
];
var ListCostCategoryDefinitions = [
    9,
    n0,
    _LCCD,
    0,
    () => ListCostCategoryDefinitionsRequest,
    () => ListCostCategoryDefinitionsResponse,
];
var ListSavingsPlansPurchaseRecommendationGeneration = [
    9,
    n0,
    _LSPPRG,
    0,
    () => ListSavingsPlansPurchaseRecommendationGenerationRequest,
    () => ListSavingsPlansPurchaseRecommendationGenerationResponse,
];
var ListTagsForResource = [
    9,
    n0,
    _LTFR,
    0,
    () => ListTagsForResourceRequest,
    () => ListTagsForResourceResponse,
];
var ProvideAnomalyFeedback = [
    9,
    n0,
    _PAF,
    0,
    () => ProvideAnomalyFeedbackRequest,
    () => ProvideAnomalyFeedbackResponse,
];
var StartCommitmentPurchaseAnalysis = [
    9,
    n0,
    _SCPA,
    0,
    () => StartCommitmentPurchaseAnalysisRequest,
    () => StartCommitmentPurchaseAnalysisResponse,
];
var StartCostAllocationTagBackfill = [
    9,
    n0,
    _SCATB,
    0,
    () => StartCostAllocationTagBackfillRequest,
    () => StartCostAllocationTagBackfillResponse,
];
var StartSavingsPlansPurchaseRecommendationGeneration = [
    9,
    n0,
    _SSPPRG,
    0,
    () => StartSavingsPlansPurchaseRecommendationGenerationRequest,
    () => StartSavingsPlansPurchaseRecommendationGenerationResponse,
];
var TagResource = [9, n0, _TRa, 0, () => TagResourceRequest, () => TagResourceResponse];
var UntagResource = [
    9,
    n0,
    _UR,
    0,
    () => UntagResourceRequest,
    () => UntagResourceResponse,
];
var UpdateAnomalyMonitor = [
    9,
    n0,
    _UAM,
    0,
    () => UpdateAnomalyMonitorRequest,
    () => UpdateAnomalyMonitorResponse,
];
var UpdateAnomalySubscription = [
    9,
    n0,
    _UAS,
    0,
    () => UpdateAnomalySubscriptionRequest,
    () => UpdateAnomalySubscriptionResponse,
];
var UpdateCostAllocationTagsStatus = [
    9,
    n0,
    _UCATS,
    0,
    () => UpdateCostAllocationTagsStatusRequest,
    () => UpdateCostAllocationTagsStatusResponse,
];
var UpdateCostCategoryDefinition = [
    9,
    n0,
    _UCCD,
    0,
    () => UpdateCostCategoryDefinitionRequest,
    () => UpdateCostCategoryDefinitionResponse,
];

class CreateAnomalyMonitorCommand extends smithyClient.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [middlewareEndpoint.getEndpointPlugin(config, Command.getEndpointParameterInstructions())];
})
    .s("AWSInsightsIndexService", "CreateAnomalyMonitor", {})
    .n("CostExplorerClient", "CreateAnomalyMonitorCommand")
    .sc(CreateAnomalyMonitor)
    .build() {
}

class CreateAnomalySubscriptionCommand extends smithyClient.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [middlewareEndpoint.getEndpointPlugin(config, Command.getEndpointParameterInstructions())];
})
    .s("AWSInsightsIndexService", "CreateAnomalySubscription", {})
    .n("CostExplorerClient", "CreateAnomalySubscriptionCommand")
    .sc(CreateAnomalySubscription)
    .build() {
}

class CreateCostCategoryDefinitionCommand extends smithyClient.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [middlewareEndpoint.getEndpointPlugin(config, Command.getEndpointParameterInstructions())];
})
    .s("AWSInsightsIndexService", "CreateCostCategoryDefinition", {})
    .n("CostExplorerClient", "CreateCostCategoryDefinitionCommand")
    .sc(CreateCostCategoryDefinition)
    .build() {
}

class DeleteAnomalyMonitorCommand extends smithyClient.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [middlewareEndpoint.getEndpointPlugin(config, Command.getEndpointParameterInstructions())];
})
    .s("AWSInsightsIndexService", "DeleteAnomalyMonitor", {})
    .n("CostExplorerClient", "DeleteAnomalyMonitorCommand")
    .sc(DeleteAnomalyMonitor)
    .build() {
}

class DeleteAnomalySubscriptionCommand extends smithyClient.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [middlewareEndpoint.getEndpointPlugin(config, Command.getEndpointParameterInstructions())];
})
    .s("AWSInsightsIndexService", "DeleteAnomalySubscription", {})
    .n("CostExplorerClient", "DeleteAnomalySubscriptionCommand")
    .sc(DeleteAnomalySubscription)
    .build() {
}

class DeleteCostCategoryDefinitionCommand extends smithyClient.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [middlewareEndpoint.getEndpointPlugin(config, Command.getEndpointParameterInstructions())];
})
    .s("AWSInsightsIndexService", "DeleteCostCategoryDefinition", {})
    .n("CostExplorerClient", "DeleteCostCategoryDefinitionCommand")
    .sc(DeleteCostCategoryDefinition)
    .build() {
}

class DescribeCostCategoryDefinitionCommand extends smithyClient.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [middlewareEndpoint.getEndpointPlugin(config, Command.getEndpointParameterInstructions())];
})
    .s("AWSInsightsIndexService", "DescribeCostCategoryDefinition", {})
    .n("CostExplorerClient", "DescribeCostCategoryDefinitionCommand")
    .sc(DescribeCostCategoryDefinition)
    .build() {
}

class GetAnomaliesCommand extends smithyClient.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [middlewareEndpoint.getEndpointPlugin(config, Command.getEndpointParameterInstructions())];
})
    .s("AWSInsightsIndexService", "GetAnomalies", {})
    .n("CostExplorerClient", "GetAnomaliesCommand")
    .sc(GetAnomalies)
    .build() {
}

class GetAnomalyMonitorsCommand extends smithyClient.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [middlewareEndpoint.getEndpointPlugin(config, Command.getEndpointParameterInstructions())];
})
    .s("AWSInsightsIndexService", "GetAnomalyMonitors", {})
    .n("CostExplorerClient", "GetAnomalyMonitorsCommand")
    .sc(GetAnomalyMonitors)
    .build() {
}

class GetAnomalySubscriptionsCommand extends smithyClient.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [middlewareEndpoint.getEndpointPlugin(config, Command.getEndpointParameterInstructions())];
})
    .s("AWSInsightsIndexService", "GetAnomalySubscriptions", {})
    .n("CostExplorerClient", "GetAnomalySubscriptionsCommand")
    .sc(GetAnomalySubscriptions)
    .build() {
}

class GetApproximateUsageRecordsCommand extends smithyClient.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [middlewareEndpoint.getEndpointPlugin(config, Command.getEndpointParameterInstructions())];
})
    .s("AWSInsightsIndexService", "GetApproximateUsageRecords", {})
    .n("CostExplorerClient", "GetApproximateUsageRecordsCommand")
    .sc(GetApproximateUsageRecords)
    .build() {
}

class GetCommitmentPurchaseAnalysisCommand extends smithyClient.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [middlewareEndpoint.getEndpointPlugin(config, Command.getEndpointParameterInstructions())];
})
    .s("AWSInsightsIndexService", "GetCommitmentPurchaseAnalysis", {})
    .n("CostExplorerClient", "GetCommitmentPurchaseAnalysisCommand")
    .sc(GetCommitmentPurchaseAnalysis)
    .build() {
}

class GetCostAndUsageCommand extends smithyClient.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [middlewareEndpoint.getEndpointPlugin(config, Command.getEndpointParameterInstructions())];
})
    .s("AWSInsightsIndexService", "GetCostAndUsage", {})
    .n("CostExplorerClient", "GetCostAndUsageCommand")
    .sc(GetCostAndUsage)
    .build() {
}

class GetCostAndUsageComparisonsCommand extends smithyClient.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [middlewareEndpoint.getEndpointPlugin(config, Command.getEndpointParameterInstructions())];
})
    .s("AWSInsightsIndexService", "GetCostAndUsageComparisons", {})
    .n("CostExplorerClient", "GetCostAndUsageComparisonsCommand")
    .sc(GetCostAndUsageComparisons)
    .build() {
}

class GetCostAndUsageWithResourcesCommand extends smithyClient.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [middlewareEndpoint.getEndpointPlugin(config, Command.getEndpointParameterInstructions())];
})
    .s("AWSInsightsIndexService", "GetCostAndUsageWithResources", {})
    .n("CostExplorerClient", "GetCostAndUsageWithResourcesCommand")
    .sc(GetCostAndUsageWithResources)
    .build() {
}

class GetCostCategoriesCommand extends smithyClient.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [middlewareEndpoint.getEndpointPlugin(config, Command.getEndpointParameterInstructions())];
})
    .s("AWSInsightsIndexService", "GetCostCategories", {})
    .n("CostExplorerClient", "GetCostCategoriesCommand")
    .sc(GetCostCategories)
    .build() {
}

class GetCostComparisonDriversCommand extends smithyClient.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [middlewareEndpoint.getEndpointPlugin(config, Command.getEndpointParameterInstructions())];
})
    .s("AWSInsightsIndexService", "GetCostComparisonDrivers", {})
    .n("CostExplorerClient", "GetCostComparisonDriversCommand")
    .sc(GetCostComparisonDrivers)
    .build() {
}

class GetCostForecastCommand extends smithyClient.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [middlewareEndpoint.getEndpointPlugin(config, Command.getEndpointParameterInstructions())];
})
    .s("AWSInsightsIndexService", "GetCostForecast", {})
    .n("CostExplorerClient", "GetCostForecastCommand")
    .sc(GetCostForecast)
    .build() {
}

class GetDimensionValuesCommand extends smithyClient.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [middlewareEndpoint.getEndpointPlugin(config, Command.getEndpointParameterInstructions())];
})
    .s("AWSInsightsIndexService", "GetDimensionValues", {})
    .n("CostExplorerClient", "GetDimensionValuesCommand")
    .sc(GetDimensionValues)
    .build() {
}

class GetReservationCoverageCommand extends smithyClient.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [middlewareEndpoint.getEndpointPlugin(config, Command.getEndpointParameterInstructions())];
})
    .s("AWSInsightsIndexService", "GetReservationCoverage", {})
    .n("CostExplorerClient", "GetReservationCoverageCommand")
    .sc(GetReservationCoverage)
    .build() {
}

class GetReservationPurchaseRecommendationCommand extends smithyClient.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [middlewareEndpoint.getEndpointPlugin(config, Command.getEndpointParameterInstructions())];
})
    .s("AWSInsightsIndexService", "GetReservationPurchaseRecommendation", {})
    .n("CostExplorerClient", "GetReservationPurchaseRecommendationCommand")
    .sc(GetReservationPurchaseRecommendation)
    .build() {
}

class GetReservationUtilizationCommand extends smithyClient.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [middlewareEndpoint.getEndpointPlugin(config, Command.getEndpointParameterInstructions())];
})
    .s("AWSInsightsIndexService", "GetReservationUtilization", {})
    .n("CostExplorerClient", "GetReservationUtilizationCommand")
    .sc(GetReservationUtilization)
    .build() {
}

class GetRightsizingRecommendationCommand extends smithyClient.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [middlewareEndpoint.getEndpointPlugin(config, Command.getEndpointParameterInstructions())];
})
    .s("AWSInsightsIndexService", "GetRightsizingRecommendation", {})
    .n("CostExplorerClient", "GetRightsizingRecommendationCommand")
    .sc(GetRightsizingRecommendation)
    .build() {
}

class GetSavingsPlanPurchaseRecommendationDetailsCommand extends smithyClient.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [middlewareEndpoint.getEndpointPlugin(config, Command.getEndpointParameterInstructions())];
})
    .s("AWSInsightsIndexService", "GetSavingsPlanPurchaseRecommendationDetails", {})
    .n("CostExplorerClient", "GetSavingsPlanPurchaseRecommendationDetailsCommand")
    .sc(GetSavingsPlanPurchaseRecommendationDetails)
    .build() {
}

class GetSavingsPlansCoverageCommand extends smithyClient.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [middlewareEndpoint.getEndpointPlugin(config, Command.getEndpointParameterInstructions())];
})
    .s("AWSInsightsIndexService", "GetSavingsPlansCoverage", {})
    .n("CostExplorerClient", "GetSavingsPlansCoverageCommand")
    .sc(GetSavingsPlansCoverage)
    .build() {
}

class GetSavingsPlansPurchaseRecommendationCommand extends smithyClient.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [middlewareEndpoint.getEndpointPlugin(config, Command.getEndpointParameterInstructions())];
})
    .s("AWSInsightsIndexService", "GetSavingsPlansPurchaseRecommendation", {})
    .n("CostExplorerClient", "GetSavingsPlansPurchaseRecommendationCommand")
    .sc(GetSavingsPlansPurchaseRecommendation)
    .build() {
}

class GetSavingsPlansUtilizationCommand extends smithyClient.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [middlewareEndpoint.getEndpointPlugin(config, Command.getEndpointParameterInstructions())];
})
    .s("AWSInsightsIndexService", "GetSavingsPlansUtilization", {})
    .n("CostExplorerClient", "GetSavingsPlansUtilizationCommand")
    .sc(GetSavingsPlansUtilization)
    .build() {
}

class GetSavingsPlansUtilizationDetailsCommand extends smithyClient.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [middlewareEndpoint.getEndpointPlugin(config, Command.getEndpointParameterInstructions())];
})
    .s("AWSInsightsIndexService", "GetSavingsPlansUtilizationDetails", {})
    .n("CostExplorerClient", "GetSavingsPlansUtilizationDetailsCommand")
    .sc(GetSavingsPlansUtilizationDetails)
    .build() {
}

class GetTagsCommand extends smithyClient.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [middlewareEndpoint.getEndpointPlugin(config, Command.getEndpointParameterInstructions())];
})
    .s("AWSInsightsIndexService", "GetTags", {})
    .n("CostExplorerClient", "GetTagsCommand")
    .sc(GetTags)
    .build() {
}

class GetUsageForecastCommand extends smithyClient.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [middlewareEndpoint.getEndpointPlugin(config, Command.getEndpointParameterInstructions())];
})
    .s("AWSInsightsIndexService", "GetUsageForecast", {})
    .n("CostExplorerClient", "GetUsageForecastCommand")
    .sc(GetUsageForecast)
    .build() {
}

class ListCommitmentPurchaseAnalysesCommand extends smithyClient.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [middlewareEndpoint.getEndpointPlugin(config, Command.getEndpointParameterInstructions())];
})
    .s("AWSInsightsIndexService", "ListCommitmentPurchaseAnalyses", {})
    .n("CostExplorerClient", "ListCommitmentPurchaseAnalysesCommand")
    .sc(ListCommitmentPurchaseAnalyses)
    .build() {
}

class ListCostAllocationTagBackfillHistoryCommand extends smithyClient.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [middlewareEndpoint.getEndpointPlugin(config, Command.getEndpointParameterInstructions())];
})
    .s("AWSInsightsIndexService", "ListCostAllocationTagBackfillHistory", {})
    .n("CostExplorerClient", "ListCostAllocationTagBackfillHistoryCommand")
    .sc(ListCostAllocationTagBackfillHistory)
    .build() {
}

class ListCostAllocationTagsCommand extends smithyClient.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [middlewareEndpoint.getEndpointPlugin(config, Command.getEndpointParameterInstructions())];
})
    .s("AWSInsightsIndexService", "ListCostAllocationTags", {})
    .n("CostExplorerClient", "ListCostAllocationTagsCommand")
    .sc(ListCostAllocationTags)
    .build() {
}

class ListCostCategoryDefinitionsCommand extends smithyClient.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [middlewareEndpoint.getEndpointPlugin(config, Command.getEndpointParameterInstructions())];
})
    .s("AWSInsightsIndexService", "ListCostCategoryDefinitions", {})
    .n("CostExplorerClient", "ListCostCategoryDefinitionsCommand")
    .sc(ListCostCategoryDefinitions)
    .build() {
}

class ListSavingsPlansPurchaseRecommendationGenerationCommand extends smithyClient.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [middlewareEndpoint.getEndpointPlugin(config, Command.getEndpointParameterInstructions())];
})
    .s("AWSInsightsIndexService", "ListSavingsPlansPurchaseRecommendationGeneration", {})
    .n("CostExplorerClient", "ListSavingsPlansPurchaseRecommendationGenerationCommand")
    .sc(ListSavingsPlansPurchaseRecommendationGeneration)
    .build() {
}

class ListTagsForResourceCommand extends smithyClient.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [middlewareEndpoint.getEndpointPlugin(config, Command.getEndpointParameterInstructions())];
})
    .s("AWSInsightsIndexService", "ListTagsForResource", {})
    .n("CostExplorerClient", "ListTagsForResourceCommand")
    .sc(ListTagsForResource)
    .build() {
}

class ProvideAnomalyFeedbackCommand extends smithyClient.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [middlewareEndpoint.getEndpointPlugin(config, Command.getEndpointParameterInstructions())];
})
    .s("AWSInsightsIndexService", "ProvideAnomalyFeedback", {})
    .n("CostExplorerClient", "ProvideAnomalyFeedbackCommand")
    .sc(ProvideAnomalyFeedback)
    .build() {
}

class StartCommitmentPurchaseAnalysisCommand extends smithyClient.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [middlewareEndpoint.getEndpointPlugin(config, Command.getEndpointParameterInstructions())];
})
    .s("AWSInsightsIndexService", "StartCommitmentPurchaseAnalysis", {})
    .n("CostExplorerClient", "StartCommitmentPurchaseAnalysisCommand")
    .sc(StartCommitmentPurchaseAnalysis)
    .build() {
}

class StartCostAllocationTagBackfillCommand extends smithyClient.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [middlewareEndpoint.getEndpointPlugin(config, Command.getEndpointParameterInstructions())];
})
    .s("AWSInsightsIndexService", "StartCostAllocationTagBackfill", {})
    .n("CostExplorerClient", "StartCostAllocationTagBackfillCommand")
    .sc(StartCostAllocationTagBackfill)
    .build() {
}

class StartSavingsPlansPurchaseRecommendationGenerationCommand extends smithyClient.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [middlewareEndpoint.getEndpointPlugin(config, Command.getEndpointParameterInstructions())];
})
    .s("AWSInsightsIndexService", "StartSavingsPlansPurchaseRecommendationGeneration", {})
    .n("CostExplorerClient", "StartSavingsPlansPurchaseRecommendationGenerationCommand")
    .sc(StartSavingsPlansPurchaseRecommendationGeneration)
    .build() {
}

class TagResourceCommand extends smithyClient.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [middlewareEndpoint.getEndpointPlugin(config, Command.getEndpointParameterInstructions())];
})
    .s("AWSInsightsIndexService", "TagResource", {})
    .n("CostExplorerClient", "TagResourceCommand")
    .sc(TagResource)
    .build() {
}

class UntagResourceCommand extends smithyClient.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [middlewareEndpoint.getEndpointPlugin(config, Command.getEndpointParameterInstructions())];
})
    .s("AWSInsightsIndexService", "UntagResource", {})
    .n("CostExplorerClient", "UntagResourceCommand")
    .sc(UntagResource)
    .build() {
}

class UpdateAnomalyMonitorCommand extends smithyClient.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [middlewareEndpoint.getEndpointPlugin(config, Command.getEndpointParameterInstructions())];
})
    .s("AWSInsightsIndexService", "UpdateAnomalyMonitor", {})
    .n("CostExplorerClient", "UpdateAnomalyMonitorCommand")
    .sc(UpdateAnomalyMonitor)
    .build() {
}

class UpdateAnomalySubscriptionCommand extends smithyClient.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [middlewareEndpoint.getEndpointPlugin(config, Command.getEndpointParameterInstructions())];
})
    .s("AWSInsightsIndexService", "UpdateAnomalySubscription", {})
    .n("CostExplorerClient", "UpdateAnomalySubscriptionCommand")
    .sc(UpdateAnomalySubscription)
    .build() {
}

class UpdateCostAllocationTagsStatusCommand extends smithyClient.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [middlewareEndpoint.getEndpointPlugin(config, Command.getEndpointParameterInstructions())];
})
    .s("AWSInsightsIndexService", "UpdateCostAllocationTagsStatus", {})
    .n("CostExplorerClient", "UpdateCostAllocationTagsStatusCommand")
    .sc(UpdateCostAllocationTagsStatus)
    .build() {
}

class UpdateCostCategoryDefinitionCommand extends smithyClient.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [middlewareEndpoint.getEndpointPlugin(config, Command.getEndpointParameterInstructions())];
})
    .s("AWSInsightsIndexService", "UpdateCostCategoryDefinition", {})
    .n("CostExplorerClient", "UpdateCostCategoryDefinitionCommand")
    .sc(UpdateCostCategoryDefinition)
    .build() {
}

const commands = {
    CreateAnomalyMonitorCommand,
    CreateAnomalySubscriptionCommand,
    CreateCostCategoryDefinitionCommand,
    DeleteAnomalyMonitorCommand,
    DeleteAnomalySubscriptionCommand,
    DeleteCostCategoryDefinitionCommand,
    DescribeCostCategoryDefinitionCommand,
    GetAnomaliesCommand,
    GetAnomalyMonitorsCommand,
    GetAnomalySubscriptionsCommand,
    GetApproximateUsageRecordsCommand,
    GetCommitmentPurchaseAnalysisCommand,
    GetCostAndUsageCommand,
    GetCostAndUsageComparisonsCommand,
    GetCostAndUsageWithResourcesCommand,
    GetCostCategoriesCommand,
    GetCostComparisonDriversCommand,
    GetCostForecastCommand,
    GetDimensionValuesCommand,
    GetReservationCoverageCommand,
    GetReservationPurchaseRecommendationCommand,
    GetReservationUtilizationCommand,
    GetRightsizingRecommendationCommand,
    GetSavingsPlanPurchaseRecommendationDetailsCommand,
    GetSavingsPlansCoverageCommand,
    GetSavingsPlansPurchaseRecommendationCommand,
    GetSavingsPlansUtilizationCommand,
    GetSavingsPlansUtilizationDetailsCommand,
    GetTagsCommand,
    GetUsageForecastCommand,
    ListCommitmentPurchaseAnalysesCommand,
    ListCostAllocationTagBackfillHistoryCommand,
    ListCostAllocationTagsCommand,
    ListCostCategoryDefinitionsCommand,
    ListSavingsPlansPurchaseRecommendationGenerationCommand,
    ListTagsForResourceCommand,
    ProvideAnomalyFeedbackCommand,
    StartCommitmentPurchaseAnalysisCommand,
    StartCostAllocationTagBackfillCommand,
    StartSavingsPlansPurchaseRecommendationGenerationCommand,
    TagResourceCommand,
    UntagResourceCommand,
    UpdateAnomalyMonitorCommand,
    UpdateAnomalySubscriptionCommand,
    UpdateCostAllocationTagsStatusCommand,
    UpdateCostCategoryDefinitionCommand,
};
class CostExplorer extends CostExplorerClient {
}
smithyClient.createAggregatedClient(commands, CostExplorer);

const paginateGetAnomalies = core.createPaginator(CostExplorerClient, GetAnomaliesCommand, "NextPageToken", "NextPageToken", "MaxResults");

const paginateGetAnomalyMonitors = core.createPaginator(CostExplorerClient, GetAnomalyMonitorsCommand, "NextPageToken", "NextPageToken", "MaxResults");

const paginateGetAnomalySubscriptions = core.createPaginator(CostExplorerClient, GetAnomalySubscriptionsCommand, "NextPageToken", "NextPageToken", "MaxResults");

const paginateGetCostAndUsageComparisons = core.createPaginator(CostExplorerClient, GetCostAndUsageComparisonsCommand, "NextPageToken", "NextPageToken", "MaxResults");

const paginateGetCostComparisonDrivers = core.createPaginator(CostExplorerClient, GetCostComparisonDriversCommand, "NextPageToken", "NextPageToken", "MaxResults");

const paginateGetSavingsPlansCoverage = core.createPaginator(CostExplorerClient, GetSavingsPlansCoverageCommand, "NextToken", "NextToken", "MaxResults");

const paginateGetSavingsPlansUtilizationDetails = core.createPaginator(CostExplorerClient, GetSavingsPlansUtilizationDetailsCommand, "NextToken", "NextToken", "MaxResults");

const paginateListCostAllocationTagBackfillHistory = core.createPaginator(CostExplorerClient, ListCostAllocationTagBackfillHistoryCommand, "NextToken", "NextToken", "MaxResults");

const paginateListCostAllocationTags = core.createPaginator(CostExplorerClient, ListCostAllocationTagsCommand, "NextToken", "NextToken", "MaxResults");

const paginateListCostCategoryDefinitions = core.createPaginator(CostExplorerClient, ListCostCategoryDefinitionsCommand, "NextToken", "NextToken", "MaxResults");

const AccountScope = {
    LINKED: "LINKED",
    PAYER: "PAYER",
};
const AnalysisStatus = {
    FAILED: "FAILED",
    PROCESSING: "PROCESSING",
    SUCCEEDED: "SUCCEEDED",
};
const AnalysisType = {
    CUSTOM_COMMITMENT: "CUSTOM_COMMITMENT",
    MAX_SAVINGS: "MAX_SAVINGS",
};
const PaymentOption = {
    ALL_UPFRONT: "ALL_UPFRONT",
    HEAVY_UTILIZATION: "HEAVY_UTILIZATION",
    LIGHT_UTILIZATION: "LIGHT_UTILIZATION",
    MEDIUM_UTILIZATION: "MEDIUM_UTILIZATION",
    NO_UPFRONT: "NO_UPFRONT",
    PARTIAL_UPFRONT: "PARTIAL_UPFRONT",
};
const SupportedSavingsPlansType = {
    COMPUTE_SP: "COMPUTE_SP",
    DATABASE_SP: "DATABASE_SP",
    EC2_INSTANCE_SP: "EC2_INSTANCE_SP",
    SAGEMAKER_SP: "SAGEMAKER_SP",
};
const TermInYears = {
    ONE_YEAR: "ONE_YEAR",
    THREE_YEARS: "THREE_YEARS",
};
const ErrorCode = {
    INTERNAL_FAILURE: "INTERNAL_FAILURE",
    INVALID_ACCOUNT_ID: "INVALID_ACCOUNT_ID",
    INVALID_SAVINGS_PLANS_TO_ADD: "INVALID_SAVINGS_PLANS_TO_ADD",
    INVALID_SAVINGS_PLANS_TO_EXCLUDE: "INVALID_SAVINGS_PLANS_TO_EXCLUDE",
    NO_USAGE_FOUND: "NO_USAGE_FOUND",
};
const AnomalyFeedbackType = {
    NO: "NO",
    PLANNED_ACTIVITY: "PLANNED_ACTIVITY",
    YES: "YES",
};
const MonitorDimension = {
    COST_CATEGORY: "COST_CATEGORY",
    LINKED_ACCOUNT: "LINKED_ACCOUNT",
    SERVICE: "SERVICE",
    TAG: "TAG",
};
const MatchOption = {
    ABSENT: "ABSENT",
    CASE_INSENSITIVE: "CASE_INSENSITIVE",
    CASE_SENSITIVE: "CASE_SENSITIVE",
    CONTAINS: "CONTAINS",
    ENDS_WITH: "ENDS_WITH",
    EQUALS: "EQUALS",
    GREATER_THAN_OR_EQUAL: "GREATER_THAN_OR_EQUAL",
    STARTS_WITH: "STARTS_WITH",
};
const Dimension = {
    AGREEMENT_END_DATE_TIME_AFTER: "AGREEMENT_END_DATE_TIME_AFTER",
    AGREEMENT_END_DATE_TIME_BEFORE: "AGREEMENT_END_DATE_TIME_BEFORE",
    ANOMALY_TOTAL_IMPACT_ABSOLUTE: "ANOMALY_TOTAL_IMPACT_ABSOLUTE",
    ANOMALY_TOTAL_IMPACT_PERCENTAGE: "ANOMALY_TOTAL_IMPACT_PERCENTAGE",
    AZ: "AZ",
    BILLING_ENTITY: "BILLING_ENTITY",
    CACHE_ENGINE: "CACHE_ENGINE",
    DATABASE_ENGINE: "DATABASE_ENGINE",
    DEPLOYMENT_OPTION: "DEPLOYMENT_OPTION",
    INSTANCE_TYPE: "INSTANCE_TYPE",
    INSTANCE_TYPE_FAMILY: "INSTANCE_TYPE_FAMILY",
    INVOICING_ENTITY: "INVOICING_ENTITY",
    LEGAL_ENTITY_NAME: "LEGAL_ENTITY_NAME",
    LINKED_ACCOUNT: "LINKED_ACCOUNT",
    LINKED_ACCOUNT_NAME: "LINKED_ACCOUNT_NAME",
    OPERATING_SYSTEM: "OPERATING_SYSTEM",
    OPERATION: "OPERATION",
    PAYER_ACCOUNT: "PAYER_ACCOUNT",
    PAYMENT_OPTION: "PAYMENT_OPTION",
    PLATFORM: "PLATFORM",
    PURCHASE_TYPE: "PURCHASE_TYPE",
    RECORD_TYPE: "RECORD_TYPE",
    REGION: "REGION",
    RESERVATION_ID: "RESERVATION_ID",
    RESOURCE_ID: "RESOURCE_ID",
    RIGHTSIZING_TYPE: "RIGHTSIZING_TYPE",
    SAVINGS_PLANS_TYPE: "SAVINGS_PLANS_TYPE",
    SAVINGS_PLAN_ARN: "SAVINGS_PLAN_ARN",
    SCOPE: "SCOPE",
    SERVICE: "SERVICE",
    SERVICE_CODE: "SERVICE_CODE",
    SUBSCRIPTION_ID: "SUBSCRIPTION_ID",
    TENANCY: "TENANCY",
    USAGE_TYPE: "USAGE_TYPE",
    USAGE_TYPE_GROUP: "USAGE_TYPE_GROUP",
};
const MonitorType = {
    CUSTOM: "CUSTOM",
    DIMENSIONAL: "DIMENSIONAL",
};
const AnomalySubscriptionFrequency = {
    DAILY: "DAILY",
    IMMEDIATE: "IMMEDIATE",
    WEEKLY: "WEEKLY",
};
const SubscriberStatus = {
    CONFIRMED: "CONFIRMED",
    DECLINED: "DECLINED",
};
const SubscriberType = {
    EMAIL: "EMAIL",
    SNS: "SNS",
};
const ApproximationDimension = {
    RESOURCE: "RESOURCE",
    SERVICE: "SERVICE",
};
const CostCategoryInheritedValueDimensionName = {
    LINKED_ACCOUNT_NAME: "LINKED_ACCOUNT_NAME",
    TAG: "TAG",
};
const CostCategoryRuleType = {
    INHERITED_VALUE: "INHERITED_VALUE",
    REGULAR: "REGULAR",
};
const CostCategoryRuleVersion = {
    CostCategoryExpressionV1: "CostCategoryExpression.v1",
};
const CostCategorySplitChargeMethod = {
    EVEN: "EVEN",
    FIXED: "FIXED",
    PROPORTIONAL: "PROPORTIONAL",
};
const CostCategorySplitChargeRuleParameterType = {
    ALLOCATION_PERCENTAGES: "ALLOCATION_PERCENTAGES",
};
const CostCategoryStatusComponent = {
    COST_EXPLORER: "COST_EXPLORER",
};
const CostCategoryStatus = {
    APPLIED: "APPLIED",
    PROCESSING: "PROCESSING",
};
const NumericOperator = {
    BETWEEN: "BETWEEN",
    EQUAL: "EQUAL",
    GREATER_THAN: "GREATER_THAN",
    GREATER_THAN_OR_EQUAL: "GREATER_THAN_OR_EQUAL",
    LESS_THAN: "LESS_THAN",
    LESS_THAN_OR_EQUAL: "LESS_THAN_OR_EQUAL",
};
const Granularity = {
    DAILY: "DAILY",
    HOURLY: "HOURLY",
    MONTHLY: "MONTHLY",
};
const GroupDefinitionType = {
    COST_CATEGORY: "COST_CATEGORY",
    DIMENSION: "DIMENSION",
    TAG: "TAG",
};
const SortOrder = {
    ASCENDING: "ASCENDING",
    DESCENDING: "DESCENDING",
};
const Metric = {
    AMORTIZED_COST: "AMORTIZED_COST",
    BLENDED_COST: "BLENDED_COST",
    NET_AMORTIZED_COST: "NET_AMORTIZED_COST",
    NET_UNBLENDED_COST: "NET_UNBLENDED_COST",
    NORMALIZED_USAGE_AMOUNT: "NORMALIZED_USAGE_AMOUNT",
    UNBLENDED_COST: "UNBLENDED_COST",
    USAGE_QUANTITY: "USAGE_QUANTITY",
};
const Context = {
    COST_AND_USAGE: "COST_AND_USAGE",
    RESERVATIONS: "RESERVATIONS",
    SAVINGS_PLANS: "SAVINGS_PLANS",
};
const LookbackPeriodInDays = {
    SEVEN_DAYS: "SEVEN_DAYS",
    SIXTY_DAYS: "SIXTY_DAYS",
    THIRTY_DAYS: "THIRTY_DAYS",
};
const OfferingClass = {
    CONVERTIBLE: "CONVERTIBLE",
    STANDARD: "STANDARD",
};
const RecommendationTarget = {
    CROSS_INSTANCE_FAMILY: "CROSS_INSTANCE_FAMILY",
    SAME_INSTANCE_FAMILY: "SAME_INSTANCE_FAMILY",
};
const FindingReasonCode = {
    CPU_OVER_PROVISIONED: "CPU_OVER_PROVISIONED",
    CPU_UNDER_PROVISIONED: "CPU_UNDER_PROVISIONED",
    DISK_IOPS_OVER_PROVISIONED: "DISK_IOPS_OVER_PROVISIONED",
    DISK_IOPS_UNDER_PROVISIONED: "DISK_IOPS_UNDER_PROVISIONED",
    DISK_THROUGHPUT_OVER_PROVISIONED: "DISK_THROUGHPUT_OVER_PROVISIONED",
    DISK_THROUGHPUT_UNDER_PROVISIONED: "DISK_THROUGHPUT_UNDER_PROVISIONED",
    EBS_IOPS_OVER_PROVISIONED: "EBS_IOPS_OVER_PROVISIONED",
    EBS_IOPS_UNDER_PROVISIONED: "EBS_IOPS_UNDER_PROVISIONED",
    EBS_THROUGHPUT_OVER_PROVISIONED: "EBS_THROUGHPUT_OVER_PROVISIONED",
    EBS_THROUGHPUT_UNDER_PROVISIONED: "EBS_THROUGHPUT_UNDER_PROVISIONED",
    MEMORY_OVER_PROVISIONED: "MEMORY_OVER_PROVISIONED",
    MEMORY_UNDER_PROVISIONED: "MEMORY_UNDER_PROVISIONED",
    NETWORK_BANDWIDTH_OVER_PROVISIONED: "NETWORK_BANDWIDTH_OVER_PROVISIONED",
    NETWORK_BANDWIDTH_UNDER_PROVISIONED: "NETWORK_BANDWIDTH_UNDER_PROVISIONED",
    NETWORK_PPS_OVER_PROVISIONED: "NETWORK_PPS_OVER_PROVISIONED",
    NETWORK_PPS_UNDER_PROVISIONED: "NETWORK_PPS_UNDER_PROVISIONED",
};
const PlatformDifference = {
    HYPERVISOR: "HYPERVISOR",
    INSTANCE_STORE_AVAILABILITY: "INSTANCE_STORE_AVAILABILITY",
    NETWORK_INTERFACE: "NETWORK_INTERFACE",
    STORAGE_INTERFACE: "STORAGE_INTERFACE",
    VIRTUALIZATION_TYPE: "VIRTUALIZATION_TYPE",
};
const RightsizingType = {
    MODIFY: "MODIFY",
    TERMINATE: "TERMINATE",
};
const SavingsPlansDataType = {
    AMORTIZED_COMMITMENT: "AMORTIZED_COMMITMENT",
    ATTRIBUTES: "ATTRIBUTES",
    SAVINGS: "SAVINGS",
    UTILIZATION: "UTILIZATION",
};
const CostAllocationTagBackfillStatus = {
    FAILED: "FAILED",
    PROCESSING: "PROCESSING",
    SUCCEEDED: "SUCCEEDED",
};
const CostAllocationTagStatus = {
    ACTIVE: "Active",
    INACTIVE: "Inactive",
};
const CostAllocationTagType = {
    AWS_GENERATED: "AWSGenerated",
    USER_DEFINED: "UserDefined",
};
const GenerationStatus = {
    FAILED: "FAILED",
    PROCESSING: "PROCESSING",
    SUCCEEDED: "SUCCEEDED",
};

Object.defineProperty(exports, "$Command", {
    enumerable: true,
    get: function () { return smithyClient.Command; }
});
Object.defineProperty(exports, "__Client", {
    enumerable: true,
    get: function () { return smithyClient.Client; }
});
exports.AccountScope = AccountScope;
exports.AnalysisNotFoundException = AnalysisNotFoundException$1;
exports.AnalysisStatus = AnalysisStatus;
exports.AnalysisType = AnalysisType;
exports.AnomalyFeedbackType = AnomalyFeedbackType;
exports.AnomalySubscriptionFrequency = AnomalySubscriptionFrequency;
exports.ApproximationDimension = ApproximationDimension;
exports.BackfillLimitExceededException = BackfillLimitExceededException$1;
exports.BillExpirationException = BillExpirationException$1;
exports.BillingViewHealthStatusException = BillingViewHealthStatusException$1;
exports.Context = Context;
exports.CostAllocationTagBackfillStatus = CostAllocationTagBackfillStatus;
exports.CostAllocationTagStatus = CostAllocationTagStatus;
exports.CostAllocationTagType = CostAllocationTagType;
exports.CostCategoryInheritedValueDimensionName = CostCategoryInheritedValueDimensionName;
exports.CostCategoryRuleType = CostCategoryRuleType;
exports.CostCategoryRuleVersion = CostCategoryRuleVersion;
exports.CostCategorySplitChargeMethod = CostCategorySplitChargeMethod;
exports.CostCategorySplitChargeRuleParameterType = CostCategorySplitChargeRuleParameterType;
exports.CostCategoryStatus = CostCategoryStatus;
exports.CostCategoryStatusComponent = CostCategoryStatusComponent;
exports.CostExplorer = CostExplorer;
exports.CostExplorerClient = CostExplorerClient;
exports.CostExplorerServiceException = CostExplorerServiceException$1;
exports.CreateAnomalyMonitorCommand = CreateAnomalyMonitorCommand;
exports.CreateAnomalySubscriptionCommand = CreateAnomalySubscriptionCommand;
exports.CreateCostCategoryDefinitionCommand = CreateCostCategoryDefinitionCommand;
exports.DataUnavailableException = DataUnavailableException$1;
exports.DeleteAnomalyMonitorCommand = DeleteAnomalyMonitorCommand;
exports.DeleteAnomalySubscriptionCommand = DeleteAnomalySubscriptionCommand;
exports.DeleteCostCategoryDefinitionCommand = DeleteCostCategoryDefinitionCommand;
exports.DescribeCostCategoryDefinitionCommand = DescribeCostCategoryDefinitionCommand;
exports.Dimension = Dimension;
exports.ErrorCode = ErrorCode;
exports.FindingReasonCode = FindingReasonCode;
exports.GenerationExistsException = GenerationExistsException$1;
exports.GenerationStatus = GenerationStatus;
exports.GetAnomaliesCommand = GetAnomaliesCommand;
exports.GetAnomalyMonitorsCommand = GetAnomalyMonitorsCommand;
exports.GetAnomalySubscriptionsCommand = GetAnomalySubscriptionsCommand;
exports.GetApproximateUsageRecordsCommand = GetApproximateUsageRecordsCommand;
exports.GetCommitmentPurchaseAnalysisCommand = GetCommitmentPurchaseAnalysisCommand;
exports.GetCostAndUsageCommand = GetCostAndUsageCommand;
exports.GetCostAndUsageComparisonsCommand = GetCostAndUsageComparisonsCommand;
exports.GetCostAndUsageWithResourcesCommand = GetCostAndUsageWithResourcesCommand;
exports.GetCostCategoriesCommand = GetCostCategoriesCommand;
exports.GetCostComparisonDriversCommand = GetCostComparisonDriversCommand;
exports.GetCostForecastCommand = GetCostForecastCommand;
exports.GetDimensionValuesCommand = GetDimensionValuesCommand;
exports.GetReservationCoverageCommand = GetReservationCoverageCommand;
exports.GetReservationPurchaseRecommendationCommand = GetReservationPurchaseRecommendationCommand;
exports.GetReservationUtilizationCommand = GetReservationUtilizationCommand;
exports.GetRightsizingRecommendationCommand = GetRightsizingRecommendationCommand;
exports.GetSavingsPlanPurchaseRecommendationDetailsCommand = GetSavingsPlanPurchaseRecommendationDetailsCommand;
exports.GetSavingsPlansCoverageCommand = GetSavingsPlansCoverageCommand;
exports.GetSavingsPlansPurchaseRecommendationCommand = GetSavingsPlansPurchaseRecommendationCommand;
exports.GetSavingsPlansUtilizationCommand = GetSavingsPlansUtilizationCommand;
exports.GetSavingsPlansUtilizationDetailsCommand = GetSavingsPlansUtilizationDetailsCommand;
exports.GetTagsCommand = GetTagsCommand;
exports.GetUsageForecastCommand = GetUsageForecastCommand;
exports.Granularity = Granularity;
exports.GroupDefinitionType = GroupDefinitionType;
exports.InvalidNextTokenException = InvalidNextTokenException$1;
exports.LimitExceededException = LimitExceededException$1;
exports.ListCommitmentPurchaseAnalysesCommand = ListCommitmentPurchaseAnalysesCommand;
exports.ListCostAllocationTagBackfillHistoryCommand = ListCostAllocationTagBackfillHistoryCommand;
exports.ListCostAllocationTagsCommand = ListCostAllocationTagsCommand;
exports.ListCostCategoryDefinitionsCommand = ListCostCategoryDefinitionsCommand;
exports.ListSavingsPlansPurchaseRecommendationGenerationCommand = ListSavingsPlansPurchaseRecommendationGenerationCommand;
exports.ListTagsForResourceCommand = ListTagsForResourceCommand;
exports.LookbackPeriodInDays = LookbackPeriodInDays;
exports.MatchOption = MatchOption;
exports.Metric = Metric;
exports.MonitorDimension = MonitorDimension;
exports.MonitorType = MonitorType;
exports.NumericOperator = NumericOperator;
exports.OfferingClass = OfferingClass;
exports.PaymentOption = PaymentOption;
exports.PlatformDifference = PlatformDifference;
exports.ProvideAnomalyFeedbackCommand = ProvideAnomalyFeedbackCommand;
exports.RecommendationTarget = RecommendationTarget;
exports.RequestChangedException = RequestChangedException$1;
exports.ResourceNotFoundException = ResourceNotFoundException$1;
exports.RightsizingType = RightsizingType;
exports.SavingsPlansDataType = SavingsPlansDataType;
exports.ServiceQuotaExceededException = ServiceQuotaExceededException$1;
exports.SortOrder = SortOrder;
exports.StartCommitmentPurchaseAnalysisCommand = StartCommitmentPurchaseAnalysisCommand;
exports.StartCostAllocationTagBackfillCommand = StartCostAllocationTagBackfillCommand;
exports.StartSavingsPlansPurchaseRecommendationGenerationCommand = StartSavingsPlansPurchaseRecommendationGenerationCommand;
exports.SubscriberStatus = SubscriberStatus;
exports.SubscriberType = SubscriberType;
exports.SupportedSavingsPlansType = SupportedSavingsPlansType;
exports.TagResourceCommand = TagResourceCommand;
exports.TermInYears = TermInYears;
exports.TooManyTagsException = TooManyTagsException$1;
exports.UnknownMonitorException = UnknownMonitorException$1;
exports.UnknownSubscriptionException = UnknownSubscriptionException$1;
exports.UnresolvableUsageUnitException = UnresolvableUsageUnitException$1;
exports.UntagResourceCommand = UntagResourceCommand;
exports.UpdateAnomalyMonitorCommand = UpdateAnomalyMonitorCommand;
exports.UpdateAnomalySubscriptionCommand = UpdateAnomalySubscriptionCommand;
exports.UpdateCostAllocationTagsStatusCommand = UpdateCostAllocationTagsStatusCommand;
exports.UpdateCostCategoryDefinitionCommand = UpdateCostCategoryDefinitionCommand;
exports.paginateGetAnomalies = paginateGetAnomalies;
exports.paginateGetAnomalyMonitors = paginateGetAnomalyMonitors;
exports.paginateGetAnomalySubscriptions = paginateGetAnomalySubscriptions;
exports.paginateGetCostAndUsageComparisons = paginateGetCostAndUsageComparisons;
exports.paginateGetCostComparisonDrivers = paginateGetCostComparisonDrivers;
exports.paginateGetSavingsPlansCoverage = paginateGetSavingsPlansCoverage;
exports.paginateGetSavingsPlansUtilizationDetails = paginateGetSavingsPlansUtilizationDetails;
exports.paginateListCostAllocationTagBackfillHistory = paginateListCostAllocationTagBackfillHistory;
exports.paginateListCostAllocationTags = paginateListCostAllocationTags;
exports.paginateListCostCategoryDefinitions = paginateListCostCategoryDefinitions;
